---
title: "Single-cell based deconvolutions of bulk AML transciptomic samples"
author: "E Onur Karakaslar"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "~/Desktop/LUMC/")
```

```{r import libraries and utils, include=TRUE, message=FALSE, warning=FALSE}
library(MuSiC) # deconvolution method
library(xbioc) # required for MusiC to run
library(dplyr) # %>%
library(scales)
library(Seurat)
library(ggpubr)
library(ggplot2) # nice plots
library(pheatmap) # pheatmaps
library(paletteer) # nice colors
library(ggalluvial)
library(RColorBrewer) # beatiful colors
library(survminer) # survival analyses (KM-plots)
library(survival) # survival analyses (surv function)
library(SingleCellExperiment)

# save figs 
saveFigs = FALSE

# load some utility functions
source("utils/wrangleMat.R") 

# ggplot theme
theme.gg <- theme_bw(base_size = 12) + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())
```

Having established the healthy bone marrow (BM) single-cell map, we can deconvolute bulk AML patients from different cohorts (TCGA-AML, TARGET, and BEAT-AML).

Importing healthy BM map:

```{r import BM}
# loaded healthy BM map
# load("cache/scmap_BM.RDS")
ref <- readRDS("data/Integrated_TRI_VG_Seurat.RDS")


ref$label.new <- ref$predicted.celltype.l2
ref$label.new[ref$predicted.celltype.l2 %in% c("CD4 Naive", "CD8 Memory", "CD4 Effector", "MAIT", "CD4 Memory", "T Proliferating", 
                                 "CD8 Effector_2", "CD8 Naive", "CD8 Effector_1", "CD8 Effector_3")] <- "T Cells"
ref$label.new[ref$predicted.celltype.l2 %in% c("NK", "NK Proliferating", "NK CD56+", "ILC" )] <- "NK Cells"
ref$label.new[ref$predicted.celltype.l2 %in% c("transitional B","Memory B", "Naive B" )] <- "B Cells"
ref$label.new[ref$predicted.celltype.l2 %in% c("cDC1","cDC2" )] <- "cDC"

# remove the cell types with less than 100 cells (Macrophage 9, Platelet 2, Stromal 18 cells)
ref <- ref[, !ref$label.new %in% c("Macrophage", "Platelet", "Stromal")]
# 
ref <- ScaleData(ref, verbose = FALSE)
ref <- RunPCA(ref, features = VariableFeatures(ref),verbose = FALSE, npcs = 30)
ref <- RunUMAP(ref, dims = 1:30)


```

```{r create C.eset}
C.eset <- ExpressionSet(assayData = as.matrix(GetAssayData(ref)), phenoData = new("AnnotatedDataFrame",ref@meta.data))

# cell types ordered according to cell counts
cell.count.df <- plyr::count(as.data.frame(C.eset$label.new))

cell.type.order <- cell.count.df[rev(order(cell.count.df[,2])), 1]
cell.type.order
```


```{r}
pdf("output/Integrated_number_of_cells.pdf", width = 8, height = 6)
ref@meta.data%>% dplyr::count_("label.new") %>% ggplot(aes(x =reorder(label.new, n), y = n)) + 
  geom_bar(stat = "identity") + 
  scale_fill_manual(values = c("#ad001a")) + geom_text(aes(label = n), nudge_y = -0.2, color ="white", size = 3) + 
  theme.gg + labs(y = "Number of cells", x = "") + scale_y_log10() + coord_flip()
dev.off()
```



```{r control experiment}
# xmat <- ref@assays$integrated
# 
# # exponentiate the data to non-log scale
# xmat@data <- expm1(xmat@data)
# 
# xmeta <- ref@meta.data
# xmeta$Cell <- rownames(xmeta)
# 
# sample.proportions <- seq(0.5, 1, 0.1)
# total.cells <- c(1000)
# 
# 
# C.eset <- Biobase::ExpressionSet(assayData = as.matrix(xmat@data), phenoData = Biobase::AnnotatedDataFrame(xmeta))
# breaksList = seq(0,1, by = .001)
# 
# lapply(sample.proportions, function(sample.proportion){
#   lapply(total.cells,  function(total.cell){
#     
#     sim.true <-  NULL
#     sim.bulks <- lapply(cell.type.order, function(x){
#       
#       real.val <- xmeta[xmeta$label.new == x,] %>% 
#         slice_sample(n = sample.proportion * total.cell, replace = T)
#       
#       noise <- xmeta[xmeta$label.new != x,] %>% 
#         add_count(label.new) %>% 
#         mutate (n = 1/n) %>% 
#         slice_sample(n = ((1-sample.proportion) * total.cell),weight_by = n, replace = T) %>% 
#         dplyr::select(-n)
#       
#       cells <- rbind(real.val, noise)
#       
#       sim.true[[x]] <<- table(cells$label.new)/sum(table(cells$label.new))
#       
#       sim.bulk <- rowSums(xmat[, cells$Cell])
#       
#       return(sim.bulk)
#     }) %>% do.call(cbind, .) %>% as.data.frame
#     
#     colnames(sim.bulks) <- cell.type.order
#     
#     # # Simulated bulks self deconvolution
#     # sim.results <- CIBERSORT((pb.sum), (sim.bulks), perm = 1000, QN = F)
#     
#     # calculate overall rmse
# 
#     # sometimes there are minus values
#     sim.bulks[sim.bulks <0] <- 0
#     T.eset <- ExpressionSet(assayData = as.matrix(sim.bulks))
#     
#     sim.results <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
#                                                 markers = NULL, normalize = FALSE, samples = 'Sample', 
#                                                 verbose = F)
#     
#     # total.rmse <- lapply(cell.type.order, function(x){
#     #   
#     #   y <- sim.true[[x]][cell.type.order] 
#     #   yhat <- sim.results[x, cell.type.order]
#     #   
#     #   error <- y - yhat
#     #   
#     #   sum(error^2, na.rm = T)
#     # }) %>% unlist %>% sum
#     
#     music.results <- sim.results$Est.prop.weighted
#     
#     missing.cols <- cell.type.order[!cell.type.order %in% colnames(music.results)]
#     
#     new.order <- setdiff(cell.type.order, missing.cols)
#     
#     pdf(paste0("output/Integrated_pseudobulk_deconv_Music_", sample.proportion,"prop_",total.cell,"cells.pdf"), 
#         width = 12, height = 12)
#     pheatmap(music.results[new.order,new.order], cluster_rows = F, cluster_cols = F,
#              color = pheatmap.colors, breaks = breaksList, border_color = NA, display_numbers = T, fontsize = 7,
#              cellwidth = 15, cellheight = 15)
#     dev.off()
#         
#     # 
#     # breaksList = seq(0,1, by = .001)
#     # pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))
#     # pdf(paste0("output/Integrated_pseudobulk_deconv_CIBERSORT_", sample.proportion,"prop_",total.cell,"cells.pdf"), width = 12, height = 12)
#     # pheatmap(sim.results[cell.type.order, cell.type.order], cluster_cols = F, cluster_rows = F, 
#     #          color = pheatmap.colors, breaks = breaksList, border_color = NA, display_numbers = T, fontsize = 7,
#     #          cellwidth = 15, cellheight = 15)
#     # dev.off()
#     
#     # return(total.rmse)
#     
#   })
# })
```






```{r manton 40K for reference}
# install.packages("~/Downloads/hcabm40k.SeuratData_3.0.0.tar.gz", repos = NULL, type = "source")

library(hcabm40k.SeuratData)
data("hcabm40k")

hcabm40k <- UpdateSeuratObject(hcabm40k)

predictions <- read.delim('data/azimuth_pred_HCA40K.tsv', row.names = 1)
hcabm40k <- AddMetaData(
	object = hcabm40k,
	metadata = predictions)

hcabm40k$label.new <- hcabm40k$predicted.celltype.l2
hcabm40k$label.new[hcabm40k$predicted.celltype.l2 %in% c("CD4 Naive", "CD8 Memory", "CD4 Effector", "MAIT", "CD4 Memory", "T Proliferating", 
                                 "CD8 Effector_2", "CD8 Naive", "CD8 Effector_1", "CD8 Effector_3")] <- "T Cells"
hcabm40k$label.new[hcabm40k$predicted.celltype.l2 %in% c("NK", "NK Proliferating", "NK CD56+", "ILC" )] <- "NK Cells"
hcabm40k$label.new[hcabm40k$predicted.celltype.l2 %in% c("transitional B","Memory B", "Naive B" )] <- "B Cells"
hcabm40k$label.new[hcabm40k$predicted.celltype.l2 %in% c("cDC1","cDC2" )] <- "cDC"


# remove the cell types with less than 100 cells (Macrophage 9, Platelet 2, Stromal 18 cells)
hcabm40k <- hcabm40k[, !hcabm40k$label.new %in% c("Macrophage", "Platelet", "Stromal")]
# 
hcabm40k <- NormalizeData(hcabm40k)
hcabm40k <- FindVariableFeatures(object = hcabm40k)
hcabm40k <- ScaleData(hcabm40k, verbose = FALSE)
hcabm40k <- RunPCA(hcabm40k, features = VariableFeatures(hcabm40k),verbose = FALSE)
hcabm40k <- RunUMAP(hcabm40k, dims = 1:50)



pdf("output/UMAP_Manton_samples.pdf", width = 12, height = 8)

DimPlot(hcabm40k, repel = T, group.by = "label.new", label.size = 4, shuffle = T,
        cols = cell.type.colors[!names(cell.type.colors) %in% c("Mixed Profiles", "Others")]) + theme.gg + ggtitle("")
dev.off()


```


```{r}
xmeta <- ref@meta.data

xmat <- ref@assays$integrated

# # exponentiate the data to non-log scale
xmat@data <- expm1(xmat@data)


xmeta$cell <- rownames(xmeta)

cell.types.hca40k <- unique(xmeta$label.new)

sample.proportion <- 0.8
total.cell <- 1e3

sim.true <- matrix(data = 0, nrow = length(cell.types.hca40k), ncol = length(cell.type.order))

colnames(sim.true) <- cell.type.order
rownames(sim.true) <- cell.types.hca40k

sim.bulks <- lapply(cell.types.hca40k, function(x){
  
  real.val <- xmeta[xmeta$label.new == x,] %>% 
    slice_sample(n = sample.proportion * total.cell, replace = T)
  
  noise <- xmeta[xmeta$label.new != x,] %>% 
    add_count(label.new) %>% 
    mutate (n = 1/n) %>% 
    slice_sample(n = ((1-sample.proportion) * total.cell),weight_by = n, replace = T) %>% 
    dplyr::select(-n)
  
  cells <- rbind(real.val, noise)
  
  tbl <- table(cells$label.new)/sum(table(cells$label.new))
  
  sim.true[x,names(tbl)] <<- tbl
  
  sim.bulk <- rowSums(xmat[, cells$cell])
  
  return(sim.bulk)
}) %>% do.call(cbind, .) %>% as.data.frame

colnames(sim.bulks) <- cell.types.hca40k
    
# sanity
sim.bulks[sim.bulks <0] <- 0

T.eset <- ExpressionSet(assayData = as.matrix(sim.bulks))

sim.results <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample', 
                                            verbose = F)


weights <- sim.results$Est.prop.weighted


pdf("output/Music_Ref_sc_simulation_results.pdf", width = 8, height = 8)
pheatmap(weights[intersect(cell.type.order, cell.types.hca40k),intersect(cell.type.order, colnames(weights))], 
         cluster_cols = F, cluster_rows = F, color = pheatmap.colors, display_numbers = T,
         cellwidth = 15, cellheight = 15, fontsize = 7, 
         border_color = NA)
dev.off()

pdf("output/Music_Ref_sc_simulation_results_real.pdf", width = 8, height = 8)
pheatmap(sim.true[intersect(cell.type.order, cell.types.hca40k),intersect(cell.type.order, colnames(weights))], 
         cluster_cols = F, cluster_rows = F, color = pheatmap.colors, display_numbers = T,
         cellwidth = 15, cellheight = 15, fontsize = 7, 
         border_color = NA)
dev.off()
```



```{r}
xmeta <- hcabm40k@meta.data

xmat <- hcabm40k@assays$RNA

# # exponentiate the data to non-log scale
xmat@data <- expm1(xmat@data)


xmeta$cell <- rownames(xmeta)

cell.types.hca40k <- unique(xmeta$label.new)

sample.proportion <- 0.8
total.cell <- 1e3

sim.true <- matrix(data = 0, nrow = length(cell.types.hca40k), ncol = length(cell.type.order))

colnames(sim.true) <- cell.type.order
rownames(sim.true) <- cell.types.hca40k
sim.bulks <- lapply(cell.types.hca40k, function(x){
  
  real.val <- xmeta[xmeta$label.new == x,] %>% 
    slice_sample(n = sample.proportion * total.cell, replace = T)
  
  noise <- xmeta[xmeta$label.new != x,] %>% 
    add_count(label.new) %>% 
    mutate (n = 1/n) %>% 
    slice_sample(n = ((1-sample.proportion) * total.cell),weight_by = n, replace = T) %>% 
    dplyr::select(-n)
  
  cells <- rbind(real.val, noise)
  
  tbl <- table(cells$label.new)/sum(table(cells$label.new))
  
  sim.true[x,names(tbl)] <<- tbl
  
  sim.bulk <- rowSums(xmat[, cells$cell])
  
  return(sim.bulk)
}) %>% do.call(cbind, .) %>% as.data.frame

colnames(sim.bulks) <- cell.types.hca40k
    
# sanity
sim.bulks[sim.bulks <0] <- 0

T.eset <- ExpressionSet(assayData = as.matrix(sim.bulks))

sim.results <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample', 
                                            verbose = F)


weights <- sim.results$Est.prop.weighted


pdf("output/Music_hca40k_sc_simulation_results.pdf", width = 8, height = 8)
pheatmap(weights[intersect(cell.type.order, cell.types.hca40k),intersect(cell.type.order, colnames(weights))], 
         cluster_cols = F, cluster_rows = F, color = pheatmap.colors, display_numbers = T,
         cellwidth = 15, cellheight = 15, fontsize = 7, 
         border_color = NA)
dev.off()

pdf("output/Music_hca40k_sc_simulation_results_real.pdf", width = 8, height = 8)
pheatmap(sim.true[intersect(cell.type.order, cell.types.hca40k),intersect(cell.type.order, colnames(weights))], 
         cluster_cols = F, cluster_rows = F, color = pheatmap.colors, display_numbers = T,
         cellwidth = 15, cellheight = 15, fontsize = 7, 
         border_color = NA)
dev.off()
```

### TCGA

Now we can start our analyses with TCGA-AML data,

```{r prepare TCGA data, cache=TRUE}
# load count matrix 
count.TCGA <- read.csv("data/Studies_Latest/TCGA/TCGA_ht_seq.csv")

count.TCGA <- wrangleMat(count.TCGA)

meta.TCGA <- read.csv("data/Studies_Latest/TCGA/TCGA_ht_seq_meta.csv", row.names = 1)

# calculate APS and 'stemness score'
meta.TCGA$APS <- calculateAPS(count.TCGA)
meta.TCGA$Stemness <- calculateStemness(count.TCGA)

meta.TCGA[meta.TCGA$primary_diagnosis == "Mutated NPM1", "primary_diagnosis"] <- 
    "Acute myeloid leukemia with mutated NPM1"
meta.TCGA[meta.TCGA$primary_diagnosis == "PML-RARa", "primary_diagnosis"] <- 
    "Acute promyelocytic leukaemia, PML-RAR-alpha"
meta.TCGA[meta.TCGA$primary_diagnosis == "inv(16) CBF-beta/MYH11", "primary_diagnosis"] <- 
    "Acute myeloid leukemia, CBF-beta/MYH11"
meta.TCGA[meta.TCGA$primary_diagnosis == "t(8;21) RUNX1-RUNX1T1", "primary_diagnosis"] <- 
    "Acute myeloid leukemia with t(8;21)(q22;q22); RUNX1-RUNX1T1"
meta.TCGA[meta.TCGA$primary_diagnosis == "t(9;11) MLLT3-MLL", "primary_diagnosis"] <- 
    "Acute myeloid leukemia with t(9;11)(p22;q23); MLLT3-MLL"


meta.TCGA.extra <- read.csv("data/Other studies/TCGA/data_clinical_patient.txt", skip = 4, sep = "\t")
meta.TCGA.extra$RISK_MOLECULAR <- ifelse(meta.TCGA.extra$RISK_MOLECULAR == "N.D.", NA,  meta.TCGA.extra$RISK_MOLECULAR)

meta.TCGA$order <- 1:nrow(meta.TCGA)
meta.TCGA <- merge(meta.TCGA, meta.TCGA.extra[, c("PATIENT_ID","FAB", "RISK_MOLECULAR", "DFS_STATUS")], by.x = "patient_ID", by.y = "PATIENT_ID", )
# re-order after merge
meta.TCGA <- meta.TCGA[order(meta.TCGA$order),]

meta.TCGA$FAB <- ifelse(meta.TCGA$FAB == "nc", NA, meta.TCGA$FAB)

meta.TCGA <- meta.TCGA %>% mutate(Age = age_at_diagnosis/365, 
                                  Sex = gender, 
                                  Recurrence = NA,
                                  Blast = PB_blasts,
                                  Origin = ifelse(grepl("Bone", sample_type), "BM", "PB"),
                                  `Primary Diagnosis` = primary_diagnosis, 
                                  ELN = RISK_MOLECULAR, 
                                  Status = ifelse(meta.TCGA$event == "Alive", 0,1),
                                  Patient = patient_ID,
                                  Sample = sample_ID,
                                  Other.Sample = NA,
                                  Study = "TCGA") 
meta.TCGA$ELN <- factor(meta.TCGA$ELN, labels = c("Favorable", "Intermediate", "Adverse"))

# 
# meta.TCGA %>% 
#   dplyr::select(Age, Sex, `Primary Diagnosis`, FAB, ELN) %>% 
#   gtsummary::tbl_summary(statistic = list(all_categorical() ~ "{n} / {N} ({p}%)")) 



# mut.TCGA <- read.csv("data/Other studies/TCGA/TCGA_mutations_2_oh_meta.csv", row.names = 1)


# primary diagnoses
pdiagnoses.TCGA <- stats::model.matrix(~ 0 + meta.TCGA$primary_diagnosis)
colnames(pdiagnoses.TCGA) <- sub(pattern = "meta.TCGA\\$primary_diagnosis", 
                                 replacement = "", colnames(pdiagnoses.TCGA))


# ens to symbol map
ens2gene <- cinaR::grch38
m <- match(rownames(count.TCGA), ens2gene$ensgene)
mapped.genes <- ens2gene$symbol[m]

# duplicated name/NA/mitochondrial genes
removed.genes <- duplicated(mapped.genes) | is.na(mapped.genes) | grepl("^MT", mapped.genes)

count.TCGA <- count.TCGA[!removed.genes,]
rownames(count.TCGA) <- mapped.genes[!removed.genes]
```

We have prepared the meta / count matrix of TCGA cohort for down-stream analyses.

```{r deconvolute TCGA, cache=TRUE, cache.vars='deconv.TCGA'}
T.eset <- ExpressionSet(assayData = as.matrix(count.TCGA))

# MusiC deconvolution
deconv.TCGA <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample', 
                                            verbose = F)$Est.prop.weighted

# deconvoluted fractions of TCGA samples (first 3 cell types)
head(deconv.TCGA)[,1:3]
```

Now that we have the deconvoluted fractions, we can create a heatmap with patients on the rows and cell types on the column, and have a look at it but first, we'll assign colors to each deconvoluted cell type for results' reproducibility.

```{r determine colors for each cell type}
# cell.type.colors <- c(as.vector(paletteer::paletteer_d("ggthemes::Tableau_20")), "#7d1a1a", "#595959")
cell.type.colors <- DiscretePalette(ncol(deconv.TCGA))
cell.type.colors <- c(cell.type.colors, "#595959")
names(cell.type.colors) <- (c(colnames(deconv.TCGA), "Others"))
```

```{r}
  
pdf("output/UMAP_Integrated_cell_type_labeled_azimuth.pdf", width = 8, height = 6)
DimPlot(ref, repel = T, group.by = "predicted.celltype.l2",pt.size = 0.3,
        label.size = 4, shuffle = T, cols = DiscretePalette(40)) + coord_fixed() + theme.gg + ggtitle("")
dev.off()

pdf("output/UMAP_Integrated_cell_type_labeled_azimuth_refined.pdf", width = 8, height = 6)
DimPlot(ref, repel = T, group.by = "label.new", label.size = 4, shuffle = T,
        cols = cell.type.colors[!names(cell.type.colors) %in% c("Mixed Profiles", "Others")]) + 
  coord_fixed() + theme.gg + ggtitle("")
dev.off()


pdf("output/UMAP_Integrated_Sample_labeled_azimuth_refined.pdf", width = 10, height = 8)
DimPlot(ref, group.by = "Sample", label.size = 4, shuffle = T,pt.size = 0.1,
        cols = paletteer_d("ggsci::category20_d3")) +  
  coord_fixed() + theme.gg + ggtitle("")
dev.off()

pdf("output/UMAP_Integrated_Study_labeled_azimuth_refined.pdf", width = 10, height = 8)
DimPlot(ref, group.by = "study", label.size = 4, shuffle = T,pt.size = 0.1,
        cols = (paletteer_d("rcartocolor::Pastel"))) +  
  coord_fixed() + theme.gg + ggtitle("")
dev.off()
```

Now, we can have our heatmap:

```{r creating the immune composition heatmap, fig.width=15, fig.height=5}


# TCGA patient annotations 
ann.row.TCGA <- cbind.data.frame(pdiagnoses.TCGA, 
                                    #Blast.BM = meta.TCGA[,"BM_blasts"],
                                    Blast = meta.TCGA[,"PB_blasts"], # All samples are from PB
                                    FAB = meta.TCGA[,"FAB"],
                                    ELN = meta.TCGA[,"RISK_MOLECULAR"],
                                    Stemness = meta.TCGA[,"Stemness"])


rownames(ann.row.TCGA) <- colnames(count.TCGA)


# assignment threshold 0.5
# assign each patient to a cluster if they are dominated by a cell type (more than 50%), 
# otherwise they are mixed profiles
# clusters.TCGA <- apply(deconv.TCGA > 0.5, 1, function(x){
#   ifelse(length(x[x]) > 0, names(x[x]), "Mixed Profiles")
# })

clusters.TCGA <- apply(deconv.TCGA, 1, function(x){names(x)[which(x == max(x))]})

# assigning colors for reproducibility 
ELN.colors = c("Good" = "#00b533", "Intermediate"= "#d6d300", "Poor" = "#D5232F")
FAB.colors = paletteer_d("awtools::a_palette")
names(FAB.colors) <- paste0("M", 0:7)
ann.colors <- list(Assignment = cell.type.colors[ unique(clusters.TCGA)], 
                   ELN = ELN.colors, FAB = FAB.colors)

# order patients
patientOrder <- function(r, o, cell.type.order){
  res <- 
    lapply(intersect(c("Mixed Profiles", cell.type.order) , unique(o)), function(x){
      r.new <- as.data.frame(r[o %in% x,,drop = F])
      
      if(x != "Mixed Profiles"){
        
        r.new <- r.new[(order(r.new[,x])),]
      }
      else {
        r.new <- r.new[hclust(dist(r.new))$order,]
      }
      
      return(r.new)
    }) %>% do.call(rbind, .)
  return(rownames(res))
}

# new.order.TCGA <- names(clusters.TCGA %>% sort)
new.order.TCGA <- patientOrder(deconv.TCGA, clusters.TCGA, cell.type.order)

breaksList = seq(0,1, by = .001)
pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))


pmap.TCGA <- pheatmap(t(deconv.TCGA [new.order.TCGA, intersect(cell.type.order, colnames(deconv.TCGA))]), 
         cluster_cols = F, cluster_rows = F, 
         annotation_col = cbind(ann.row.TCGA[new.order.TCGA,], Assignment = clusters.TCGA[new.order.TCGA]),
         show_rownames = T, show_colnames = F, 
         color = pheatmap.colors, annotation_colors = ann.colors, 
         breaks = breaksList, border_color = NA,
         gaps_col = cumsum(table(clusters.TCGA[new.order.TCGA])[unique(clusters.TCGA[new.order.TCGA])]))

pmap.TCGA

```

```{r save TCGA deconv hm, include=FALSE}
if (saveFigs){
    pdf("output/Music_TCGA_deconvolution_via_integrated_pseudobulk_simple_clustering.pdf", width = 16, height = 7)
    print(pmap.TCGA)
    dev.off()
}
```

Having the cluster assignments, we can check the survival curves:

```{r survival all clusters, fig.height=4}
temp.TCGA <- cbind(meta.TCGA, cluster = clusters.TCGA)


# simple clustering
km.tcga.cluster <- survfit(Surv(overal_survival, Status) ~ cluster, data=temp.TCGA)

s
plot.surv.all.TCGA <- 
    ggsurvplot(km.tcga.cluster, 
            pval = T,
          
           # Change legends: title & labels
           legend.title = "Cluster",
           legend.labs = gsub("cluster=", "", names(km.tcga.cluster$strata)), 
          
           palette = cell.type.colors[gsub("cluster=", "", names(km.tcga.cluster$strata))],
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

plot.surv.all.TCGA
```

```{r include=FALSE}
if(saveFigs)
{    
    pdf("output/TCGA_Kaplan_Meier_simple_clustering_all.pdf")
    print(plot.surv.all.TCGA, newpage = F)
    dev.off()
}
```

```{r survival redefined clusters}

# redefine groups <15 patients as 'Others'
temp.TCGA <- temp.TCGA %>%  dplyr::add_count(cluster)
temp.TCGA$cluster.refined <- ifelse(temp.TCGA$n > 20, temp.TCGA$cluster, NA)

km.tcga.cluster.refined <- survfit(Surv(overal_survival, Status) ~ cluster.refined, data= temp.TCGA)


plot.surv.redefined.TCGA <- 
    ggsurvplot(km.tcga.cluster.refined, 
            pval = T,conf.int = F, 
          
           # Change legends: title & labels
           legend.title = "Cluster",
           legend.labs = gsub("cluster.refined=", "", names(km.tcga.cluster.refined$strata)), 
          
           palette = cell.type.colors[gsub("cluster.refined=", "", names(km.tcga.cluster.refined$strata))],
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

plot.surv.redefined.TCGA
```

```{r include=FALSE}
if(saveFigs)
{    
    pdf("output/TCGA_Kaplan_Meier_simple_clustering.pdf")
    print(plot.surv.redefined.TCGA, newpage = F)
    dev.off()
}
```

```{r}
# make GMP the reference
temp.TCGA$cluster.refined <- factor(temp.TCGA$cluster.refined, levels = c("GMP", "CD14 Mono", "HSC"))
TCGA.cox <- coxph(Surv(overal_survival, Status) ~ cluster.refined, data = temp.TCGA)
TCGA.cox
ggforest(TCGA.cox)
```

```{r}
TCGA.cox.ELN <- coxph(Surv(overal_survival, Status) ~ ELN, data = temp.TCGA)
ggforest(TCGA.cox.ELN)
```

```{r}
# ELN status as covariate
temp.TCGA$Age <- ifelse(temp.TCGA$age_at_diagnosis/365 >= 55, ">=55", "<55")
TCGA.cox2 <- coxph(Surv(overal_survival, Status) ~  ELN + Age + cluster.refined, data = temp.TCGA)
TCGA.cox2
```

```{r}
ggforest(TCGA.cox2)
```

```{r TCGA ELN CLASSES}
# among cd14, hsc, and gmp patients
km.tcga.eln <- survfit(Surv(overal_survival, Status) ~ ELN, data=temp.TCGA)
km.tcga.eln

plot.ELN.TCGA <- 
    ggsurvplot(km.tcga.eln,
           pval = T,
           # surv.median.line = "hv", # Add medians survival
          
           # Change legends: title & labels
          legend.title = "Cluster", 
          legend.labs = gsub("ELN=", "", names(km.tcga.eln$strata)),
           
           palette =  c("Favorable" = "#33CC00FF", "Intermediate" = "#CC9900FF", "Adverse" = "#991A00FF"), 
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

plot.ELN.TCGA
```

```{r include=FALSE}

if(saveFigs){
    pdf("output/TCGA_ELN_KM.pdf")
    print(plot.ELN.TCGA, newpage = F)
    dev.off()
}
```

```{r}
# make GMP the reference

rownames(temp.TCGA) <- temp.TCGA$sample_ID
temp.TCGA$cluster.refined <- factor(temp.TCGA$cluster.refined, levels = c("GMP", "CD14 Mono", "HSC"))
temp.TCGA <- as.data.frame(cbind(temp.TCGA, deconv.TCGA[,c("GMP", "CD14 Mono", "HSC")]))

cox.tcga.deconv <- coxph(Surv(overal_survival, Status) ~ GMP + `CD14 Mono` + HSC + ELN, data = temp.TCGA)
cox.tcga.deconv
X.train <- data.frame(deconv.TCGA[,c("GMP", "CD14 Mono", "HSC")], ELN = temp.TCGA$ELN, Age = temp.TCGA$Age, check.names = F)

preds <- predict(cox.tcga.deconv, newdata = X.train)

qtiles <- quantile(preds, c(0, 0.33, 0.66, 1), na.rm = T)

patient.groups.TCGA <- cut(preds, qtiles, labels = c("Low Risk", "Medium Risk", "High Risk"), include.lowest = T)
temp.TCGA$patient.groups <- patient.groups.TCGA

km.tcga.qtile <- survfit(Surv(overal_survival, Status) ~ patient.groups, data = temp.TCGA)
km.tcga.qtile

plot.tcga.ip.eln <- ggsurvplot(km.tcga.qtile,           
           pval = TRUE, # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("patient.groups=", "", names(km.tcga.qtile$strata)),
           
           palette = c("Low Risk" = "#33CC00FF", "Medium Risk" = "#CC9900FF", "High Risk" = "#991A00FF"), 
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))
plot.tcga.ip.eln
```

```{r}
df.rs <- data.frame(preds, ELN = temp.TCGA$ELN)
ggplot(df.rs, aes(preds)) + geom_histogram(bins = 50) + theme_bw() + xlab("Risk Score (binned)")
```

```{r}
library(ggridges)
ggplot(df.rs %>% filter(!is.na(ELN)), aes(preds,ELN,  fill = ELN)) + 
    geom_density_ridges(alpha = 0.5,jittered_points = TRUE, point_alpha=1,point_shape=21) + 
    theme_bw() + labs(fill = "ELN") + xlab("Risk Score") + ylab("ELN") + geom_vline(xintercept = c(qtiles[2:3]), alpha = 0.5)
```

```{r}
if(saveFigs){
  pdf("output/TCGA_IP_ELN.pdf")
  print(plot.tcga.ip.eln, newpage = F)
  dev.off()
}
```

```{r}
plot.tcga.alluvial <- temp.TCGA %>% dplyr::select(ELN, patient.groups) %>% dplyr::count(ELN, patient.groups, name = "Freq") %>% 
ggplot(aes(axis1 = ELN, axis2 = patient.groups, y = Freq)) +
  geom_flow(aes(fill = ELN)) + 
  geom_stratum() + 
  geom_text(stat = "stratum", aes(label = paste0(after_stat(stratum),"\n", percent(after_stat(prop), accuracy = 0.1))), size = 3) +
  theme_void() + scale_fill_manual(values = c("Favorable" = "#33CC00FF", "Intermediate" = "#CC9900FF", "Adverse" = "#991A00FF"))
plot.tcga.alluvial
```

```{r}
if(saveFigs){
    pdf("output/TCGA_Alluvial.pdf")
    print(plot.tcga.alluvial, newpage = F)
    dev.off()
}
```

```{r NPM1 for different cluster assignments}
temp.TCGA$NPM1.cluster <- ifelse(temp.TCGA$primary_diagnosis == "Acute myeloid leukemia with mutated NPM1",
                                 ifelse(temp.TCGA$cluster == "GMP", 
                                        "NPM1 (GMP)", "NPM1 (Others)"), NA)

km.tcga.npm1 <- survfit(Surv(overal_survival, Status) ~ NPM1.cluster, data=temp.TCGA)


plot.promyelocytes.NPM1.TCGA <- 
    ggsurvplot(km.tcga.npm1, 
           pval = T,
           # surv.median.line = "hv", # Add medians survival
          
           # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("NPM1.cluster=", "", names(km.tcga.npm1$strata)),
           
           palette = "lancet",  
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

plot.promyelocytes.NPM1.TCGA
```

```{r include=FALSE}

if(saveFigs){
    pdf("output/TCGA_Pro_vs_Others_KM_for_NPM1.pdf")
    print(plot.promyelocytes.NPM1.TCGA, newpage = F)
    dev.off()
}
```

```{r NOS comparison}
temp.TCGA$NOS.cluster <- ifelse(temp.TCGA$primary_diagnosis == "Acute myeloid leukemia, NOS",
                                 ifelse(temp.TCGA$cluster == "GMP", 
                                        "AML, NOS (GMP)", "AML, NOS (Others)"), NA)


km.tcga.nos <- survfit(Surv(overal_survival, Status) ~ NOS.cluster, data=temp.TCGA[!is.na(temp.TCGA$NOS.cluster), ])


plot.promyelocytes.NOS.TCGA <- 
    ggsurvplot(km.tcga.nos, 
           pval = T,
           # surv.median.line = "hv", # Add medians survival
          
           # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("NOS.cluster=", "", names(km.tcga.nos$strata)),
           
           palette = "lancet",  
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

plot.promyelocytes.NOS.TCGA
```

```{r}
if(saveFigs){
    pdf("output/TCGA_GMP_vs_Others_KM_for_NOS.pdf")
    print(plot.promyelocytes.NOS.TCGA, newpage = F)
    dev.off()
}
```

We can also project the [stemness score](https://www.nature.com/articles/nature20598) of the patients per cluster.

```{r}

p1 <- ggplot(temp.TCGA, aes(cluster, Stemness, fill = cluster)) + 
    geom_boxplot() + geom_jitter(width = 0.1) + 
  theme.gg + ylab("Stemness Score") + xlab("") + ggtitle("TCGA") + 
  scale_fill_manual(values = cell.type.colors[unique(temp.TCGA$cluster)])+ 
  labs(fill = "Clusters") + theme(axis.text.x=element_text(angle=45,hjust=1)) 
  
p2 <- ggplot(temp.TCGA, aes(cluster.refined, Stemness, fill = cluster.refined)) + 
    geom_boxplot() + geom_jitter(width = 0.1) + 
  theme.gg + ylab("Stemness Score") + xlab("") + ggtitle("TCGA") + 
  scale_fill_manual(values = cell.type.colors[unique(temp.TCGA$cluster.refined)])+ 
  labs(fill = "Refined Clusters") + theme(axis.text.x=element_text(angle=45,hjust=1)) 


print(p1)

print(p2)
```

```{r include=FALSE}
if(saveFigs){
    pdf("output/TCGA_Stemness_vs_Refined_Clusters.pdf", width = 5, height = 5)
    p1 
    p2
    dev.off()
}
```

### LUMC

We have 100 bulk RNA-seq samples from LUMC (in house data). We can deconvolute them using the same pipeline.

```{r deconvoluting LUMC, cache=TRUE}
# count.TCGA <- data.table::fread("data/Studies_Latest/TCGA/TCGA_ht_seq.csv")
# 
# count.TCGA <- as.data.frame(t(count.TCGA))
# 
# colnames(count.TCGA) <- count.TCGA[1,]
# count.TCGA <- count.TCGA[-1,]
# 
# write.csv(count.TCGA, "data/Studies_Latest/TCGA/TCGA_ht_seq.csv")


# count matrix
count.LUMC <- read.csv("data/Studies_Latest/LUMC-AML/LUMC_counts_htseq.csv")

count.LUMC <- wrangleMat(count.LUMC)

meta.LUMC <- read.csv("data/Studies_Latest/LUMC-AML/meta_AML_LUMC.csv", row.names = 1)

# meta.LUMC.jeppe <- read.csv("~/Desktop/LUMC_Deconvolution_with_Jeppe_clusters.csv")
# meta.LUMC <- cbind(meta.LUMC, meta.LUMC.jeppe[,c("jeppe_clusters", "available_for_scRNAseq")])
meta.LUMC$APS <- calculateAPS(count.LUMC)
meta.LUMC$Stemness <- calculateStemness(count.LUMC)
# meta.LUMC.diagnosis <- read.csv("data/Other studies/LUMC/meta_2.csv", sep = ";", row.names = 1)
# meta.LUMC <- cbind(meta.LUMC, meta.LUMC.diagnosis)

# match all primary diagnosis to WHO 2016
meta.LUMC[meta.LUMC$sample_description %in% c("Acute monoblastic/monocytic leukemia",
                                              "Acute monoblastic/monocytic leukemia ",
                                              "Acute myelomonocytic leukemia",
                                              "AML NOS with maturation",
                                              "AML NOS with minimal differentiation",
                                              "AML NOS without maturation"), "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia, NOS"


meta.LUMC[meta.LUMC$sample_description %in% c("AML with inv(16), CBFB-MYH11"), "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia, CBF-beta/MYH11"

meta.LUMC[meta.LUMC$sample_description %in% c("AML with mutated NPM1"), "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with mutated NPM1"

meta.LUMC[meta.LUMC$sample_description %in% c("AML with inv(3), GATA2, MECOM"), "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with inv(3)(q21q26.2) or t(3;3)(q21;q26.2); RPN1-EVI1"


meta.LUMC[meta.LUMC$sample_description %in% c("AML with t(8,21), RUNX1-RUNX1T1"), "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(8;21)(q22;q22); RUNX1-RUNX1T1"

meta.LUMC[meta.LUMC$sample_description %in% c("AML with t(9,11), KMT2A-MLLT3"), "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(9;11)(p22;q23); MLLT3-MLL"

meta.LUMC[meta.LUMC$sample_description %in% c("Acute promyelocytic leukemia with t(15,17), PML-RARA"), "primary_diagnosis_refined"] <-
    "Acute promyelocytic leukaemia, PML-RAR-alpha"

meta.LUMC[meta.LUMC$sample_description %in% c("AML with myelodysplasia-related changes"), "primary_diagnosis_refined"] <-
    "Acute myeloid leukemia with myelodysplasia-related changes"


meta.LUMC[meta.LUMC$sample_description %in% c("Therapy-related AML"), "primary_diagnosis_refined"] <-
    "Therapy related myeloid neoplasm"

meta.LUMC[meta.LUMC$sample_description %in% c("AML with t(6,9), DEK-NUP214"), "primary_diagnosis_refined"] <-
    "Acute myeloid leukemia with t(6;9)(p23;q34); DEK-NUP214"


meta.LUMC <- meta.LUMC %>% mutate(Age = NA, 
                                  Sex = ifelse(gender == "F", "female", "male"), 
                                  Recurrence = NA,
                                  Origin = sample.type,
                                  Blast = blast_percentage,
                                  `Primary Diagnosis` = primary_diagnosis_refined, 
                                  ELN = NA, 
                                  FAB = NA,
                                  Status = NA,
                                  overal_survival = NA,
                                  Patient = subject_id,
                                  Sample = sample_title,
                                  Other.Sample = NA,
                                  Study = "LUMC") 

rownames(meta.LUMC) <- meta.LUMC$Sample

colnames(count.LUMC) <- rownames(meta.LUMC)
# mut.LUMC <- read.csv("data/Other studies/LUMC/LUMC_mutations_oh_meta.csv", row.names = 1)

#rownames(mut.LUMC) <- colnames(count.LUMC)

# primary diagnoses
pdiagnoses.LUMC <- stats::model.matrix(~ 0 + meta.LUMC$`Primary Diagnosis`)
colnames(pdiagnoses.LUMC) <- sub(pattern = "meta.LUMC\\$`Primary Diagnosis`", 
                                 replacement = "", colnames(pdiagnoses.LUMC))

# Union FLT3 mutations to get a vector
# mut.LUMC[,"FLT3"] <- ifelse(rowSums(mut.LUMC[,c("FLT3.ITD", "FLT3.TKD")] == 1) > 0, 1, 0)


# ens to symbol map
ens2gene <- cinaR::grch38
m <- match(rownames(count.LUMC), ens2gene$ensgene)
mapped.genes <- ens2gene$symbol[m]

# duplicated name/NA/mitochondrial genes
removed.genes <- duplicated(mapped.genes) | is.na(mapped.genes) | grepl("^MT", mapped.genes)

count.LUMC <- count.LUMC[!removed.genes,]
rownames(count.LUMC) <- mapped.genes[!removed.genes]


T.eset <- ExpressionSet(assayData = as.matrix(count.LUMC))

deconv.LUMC <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample', 
                                            verbose = F)$Est.prop.weighted
```

Now that, we have the deconvoluted fractions we can have the heatmap:

```{r heatmap LUMC, fig.width=15, fig.height=5, cache=TRUE}

ann.row.LUMC <- as.data.frame(cbind(pdiagnoses.LUMC, 
                                    Stemness = meta.LUMC[,"Stemness"] ,
                                    Blast = meta.LUMC[,"blast_percentage"]
                                    ))

rownames(ann.row.LUMC) <- colnames(count.LUMC)

# assignment threshold
# clusters.LUMC <- apply(deconv.LUMC > 0.5, 1, function(x){
#   ifelse(length(x[x]) > 0, names(x[x]), "Mixed Profiles")
# }) 
clusters.LUMC<- apply(deconv.LUMC, 1, function(x){names(x)[which(x == max(x))]})

#new.order.LUMC <- names(clusters.LUMC %>% sort)
new.order.LUMC <- patientOrder(deconv.LUMC, clusters.LUMC, cell.type.order)


# assigning colors for reproducibility 
ann.colors <- list(cluster = cell.type.colors[ unique(clusters.LUMC)])

breaksList = seq(0,1, by = .001)
pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))

pmap.LUMC <- pheatmap(t(deconv.LUMC [new.order.LUMC, intersect(cell.type.order, colnames(deconv.LUMC))]), 
         cluster_cols = F, cluster_rows = F, show_rownames = T, show_colnames = F,
         annotation_col = cbind(ann.row.LUMC[new.order.LUMC,], cluster = clusters.LUMC[new.order.LUMC]),
         color = pheatmap.colors, annotation_colors = ann.colors,
         breaks = breaksList, border_color = NA,
         gaps_col = cumsum(table(clusters.LUMC[new.order.LUMC])[unique(clusters.LUMC[new.order.LUMC])]))

pmap.LUMC

```

```{r include=FALSE}
if(saveFigs){
    pdf("output/Music_LUMC_deconvolution_via_integrated_pseudobulk_simple_clustering.pdf", width = 16, height = 5)
    pmap.LUMC
    dev.off()
}
```

```{r stemness vs refined clusters LUMC}
temp.LUMC <- cbind(meta.LUMC, cluster = clusters.LUMC)

# redefine groups <15 patients as 'Others'
temp.LUMC <- temp.LUMC %>%  dplyr::add_count(cluster)
temp.LUMC$cluster.refined <- ifelse(temp.LUMC$n > 10, temp.LUMC$cluster, "Others")

p1 <- ggplot(temp.LUMC, aes(cluster, Stemness, fill = cluster)) + 
    geom_boxplot() + geom_jitter(width = 0.1) + 
  theme.gg + ylab("Stemness Score") + xlab("") + ggtitle("LUMC") + 
  scale_fill_manual(values = cell.type.colors[unique(temp.LUMC$cluster)])+ 
  labs(fill = "Clusters") + theme(axis.text.x=element_text(angle=45,hjust=1)) 
  
p2 <- ggplot(temp.LUMC, aes(cluster.refined, Stemness, fill = cluster.refined)) + 
    geom_boxplot() + geom_jitter(width = 0.1) + 
  theme.gg + ylab("Stemness Score") + xlab("") + ggtitle("LUMC") + 
  scale_fill_manual(values = cell.type.colors[unique(temp.LUMC$cluster.refined)])+ 
  labs(fill = "Refined Clusters") + theme(axis.text.x=element_text(angle=45,hjust=1)) 


print(p1)
print(p2)

temp.LUMC <- temp.LUMC %>% add_count(cluster) %>% mutate(cluster.refined = ifelse(n>3, cluster, "Others"))

pdf("output/LUMC_Blast_vs_Cell_Types.pdf", width = 6, height = 6)
ggplot(temp.LUMC, aes(cluster.refined, Blast, fill = cluster.refined)) + 
  geom_boxplot() + geom_jitter(width = 0.1) + 
  theme.gg + ylab("Reported Blast %") + xlab("") + ggtitle("LUMC") + 
  scale_fill_manual(values = cell.type.colors)+ 
  labs(fill = "", caption = sprintf('anova p-value: %.5f', summary(aov(Blast ~cluster.refined,data= temp.LUMC))[[1]]$`Pr(>F)`[1])) + theme(axis.text.x=element_text(angle=45,hjust=1), legend.position  = "none") 
dev.off()




a <- pairwise.t.test(temp.LUMC$Blast, temp.LUMC$cluster.refined,
                 p.adjust.method = "BH")

a$p.value

writexl::write_xlsx(data.frame(a$p.value), path ="Supplementary Tables/Cell_Type_vs_Blast_ttest.xlsx")
```

```{r include=FALSE}
if(saveFigs){
    pdf("output/LUMC_Stemness_vs_Refined_Clusters.pdf", width = 5, height = 5)
    p1 
    p2
    dev.off()
}
```

### BEAT-AML

```{r BEAT AML deconvolution, cache=TRUE}
# count.BEAT <- read.csv("data/Other studies/BEAT/BEAT_AML_ht_seq.csv")
count.BEAT <- read.csv("data/Studies_Latest/BEAT-AML/BEAT_AML_ht_seq.csv")

count.BEAT <- wrangleMat(count.BEAT)

# meta.BEAT <- read.csv("data/Other studies/BEAT/BEAT_AML_ht_seq_meta.csv", row.names = 1)
meta.BEAT <- read.csv("data/Studies_Latest/BEAT-AML/BEAT_AML_ht_seq_meta_all.csv", row.names = 1)
# meta.BEAT <- read.csv("data/Studies_Latest/BEAT-AML/BEAT_AML_ht_seq_meta.csv", row.names = 1)
meta.BEAT$blasts <- ifelse(grepl("Bone Marrow",meta.BEAT$samples.sample_type), meta.BEAT$BM_blasts, meta.BEAT$PB_blasts)

meta.BEAT$APS <- calculateAPS(count.BEAT)
meta.BEAT$Stemness <- calculateStemness(count.BEAT)

#keep the order
meta.BEAT$order <- 1:nrow(meta.BEAT)
meta.BEAT$ID <- rownames(meta.BEAT)

meta.BEAT.FAB.ELN <- read.csv("data/Studies_Latest/BEAT-AML/BEAT_FAB_LABid.csv")
# meta.BEAT.ELN <- read.csv("data/Studies_Latest/BEAT-AML/BEAT_META_ELN.tsv", sep = "\t")


meta.BEAT <- merge(meta.BEAT, meta.BEAT.FAB.ELN[,c("PatientID", "LLS_PatientID", "SampleID", 
                                                   "LLS_SampleID", "FAB_BlastMorphology", "ELN2017")], 
                   by.x = "ID", by.y = "SampleID", all.x = T)




meta.BEAT <- meta.BEAT[order(meta.BEAT$order),]
rownames(meta.BEAT) <- meta.BEAT$ID

meta.BEAT$primary_diagnosis_refined <- ifelse(meta.BEAT$diagnoses.primary_diagnosis %in% c("Acute myeloid leukemia, NOS",
                                                                   "Acute myeloid leukemia with maturation",
                                                                   "Acute myeloid leukemia without maturation",
                                                                   "Acute myeloid leukemia, minimal differentiation",
                                                                   "Acute myelomonocytic leukemia",
                                                                   "Acute monoblastic and monocytic leukemia",
                                                                   "Acute erythroid leukaemia",
                                                                   "Acute megakaryoblastic leukaemia",
                                                                   "Acute basophilic leukemia",
                                                                   "Acute panmyelosis with myelofibrosis"), 
                                              "Acute myeloid leukemia, NOS", meta.BEAT$diagnoses.primary_diagnosis)

meta.BEAT$primary_diagnosis_refined <- ifelse(meta.BEAT$primary_diagnosis_refined %in% 
                                                  c("Atypical chronic myeloid leukemia, BCR/ABL negative",
                                                    "Chronic myelomonocytic leukemia, NOS"), 
                                              "MDS/MPN", meta.BEAT$primary_diagnosis_refined)


meta.BEAT$primary_diagnosis_refined <- ifelse(meta.BEAT$primary_diagnosis_refined %in% 
                                                  c("Essential thrombocythemia"), 
                                              "MPN", meta.BEAT$primary_diagnosis_refined)

meta.BEAT$primary_diagnosis_refined <- ifelse(meta.BEAT$primary_diagnosis_refined %in% 
                                                  c("Mixed phenotype acute leukemia, T/myeloid, NOS"), 
                                              "Acute leukemias of ambiguous lineage", meta.BEAT$primary_diagnosis_refined)


meta.BEAT$primary_diagnosis_refined <- ifelse(meta.BEAT$primary_diagnosis_refined %in% 
                                                  c("Myelodysplastic syndrome, unclassifiable",
                                                    "Refractory anemia with excess blasts",
                                                    "Refractory cytopenia with multilineage dysplasia"), 
                                              "MDS", meta.BEAT$primary_diagnosis_refined)


# Select AML and related neoplasms (WHO 2016 classes)
meta.BEAT<- meta.BEAT[grepl("Acute myeloid leukemia", meta.BEAT$primary_diagnosis_refined) | 
              meta.BEAT$primary_diagnosis_refined %in% c("Acute promyelocytic leukaemia, PML-RAR-alpha",
                                                         "Myeloid sarcoma",
                                                         "Therapy related myeloid neoplasm"),]

meta.BEAT <- meta.BEAT %>% mutate(Age = diagnoses.age_at_diagnosis/365, 
                                  Sex = demographic.gender, 
                                  Blast = blasts,
                                  Recurrence = ifelse(prim_rec == "Recurrent", "Yes", "No"),
                                  Origin = ifelse(grepl("Bone", samples.sample_type), "BM", "PB"),
                                  `Primary Diagnosis` = primary_diagnosis_refined, 
                                  FAB = FAB_BlastMorphology,
                                  ELN = ELN2017,
                                  Patient = meta.BEAT$case_id,
                                  Sample = rownames(.),
                                  Other.Sample = LLS_SampleID,
                                  Study = "BEAT",
                                  Status = ifelse(event == "Dead", 1, 0)) 



# We don't know it so, assigned to NA
meta.BEAT$ELN[!meta.BEAT$ELN %in% c("Favorable", "Adverse", "Intermediate")] <- NA


meta.BEAT$FAB <- ifelse(meta.BEAT$FAB %in% c("N/A", "NOS") , NA,meta.BEAT$FAB)
meta.BEAT$FAB <- sub(".*-", "", meta.BEAT$FAB)  %>% substr(1,2) 


# mut.BEAT <- read.csv("data/Other studies/BEAT/BEAT_AML_mutations.csv", row.names = 1)

# BEAT has healthy samples, filtering those from the matrices
patients.BEAT <- rownames(meta.BEAT)


count.BEAT.healthy <- count.BEAT[,setdiff(colnames(count.BEAT), patients.BEAT)]
count.BEAT <- count.BEAT[,patients.BEAT]
meta.BEAT  <- meta.BEAT [patients.BEAT,]
# mut.BEAT   <- mut.BEAT  [patients.BEAT,]


# primary diagnoses
pdiagnoses.BEAT <- stats::model.matrix(~ 0 + meta.BEAT$primary_diagnosis_refined)
colnames(pdiagnoses.BEAT) <- sub(pattern = "meta.BEAT\\$primary_diagnosis_refined",
                                 replacement = "", colnames(pdiagnoses.BEAT))


# ens to symbol map
ens2gene <- cinaR::grch38
m <- match(rownames(count.BEAT), ens2gene$ensgene)
mapped.genes <- ens2gene$symbol[m]

# duplicated name/NA/mitochondrial genes
removed.genes <- duplicated(mapped.genes) | is.na(mapped.genes) | grepl("^MT", mapped.genes)

count.BEAT <- count.BEAT[!removed.genes,rownames(meta.BEAT)]
rownames(count.BEAT) <- mapped.genes[!removed.genes]

T.eset <- ExpressionSet(assayData = as.matrix(count.BEAT))

deconv.BEAT <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample',
                                            verbose = F)$Est.prop.weighted
```

```{r cache=TRUE, fig.width=16, fig.height=5.5}
ann.row.BEAT <- as.data.frame(cbind.data.frame(pdiagnoses.BEAT, Blast = meta.BEAT$Blast, Stemness = meta.BEAT$Stemness, FAB = meta.BEAT$FAB, ELN = meta.BEAT$ELN))

rownames(ann.row.BEAT) <- colnames(count.BEAT)


# assignment threshold
clusters.BEAT <- apply(deconv.BEAT, 1, function(x){names(x)[which(x == max(x))]})




# new.order.BEAT <- names(clusters.BEAT %>% sort)
new.order.BEAT <- patientOrder(deconv.BEAT, clusters.BEAT, cell.type.order)

# some of the Blast levels are NA, imputing those using the ones that exists
df.blast <- as.data.frame(cbind(Blast = meta.BEAT$blasts, deconv.BEAT))

fit <- glm(Blast ~ ., data = df.blast, na.action = na.omit )

ann.row.BEAT$Blast.predicted <- predict(fit, newdata = df.blast[,-1])
ann.row.BEAT$Blast.predicted [ann.row.BEAT$Blast.predicted < 0]  <- 0

ann.row.BEAT$Blast.predicted <- ifelse(!is.na(ann.row.BEAT$Blast), ann.row.BEAT$Blast, ann.row.BEAT$Blast.predicted)
ann.row.BEAT <- as.data.frame(subset(ann.row.BEAT, select= -c(Blast)))



ELN.colors = c("Favorable" = "#00b533", "Intermediate"= "#d6d300", "Adverse" = "#D5232F")
FAB.colors = paletteer_d("awtools::a_palette")
names(FAB.colors) <- paste0("M", 0:7)
ann.colors <- list(cluster = cell.type.colors[ unique(clusters.BEAT)], 
                   ELN = ELN.colors, FAB = FAB.colors)

breaksList = seq(0,1, by = .001)
pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))

pmap.BEAT <- pheatmap(t(deconv.BEAT[new.order.BEAT, intersect(cell.type.order, colnames(deconv.BEAT))]),
             cluster_cols = F, cluster_rows = F, show_rownames = T, show_colnames = F,
             annotation_col = cbind(ann.row.BEAT[new.order.BEAT,], 
                                    cluster = clusters.BEAT[new.order.BEAT]),
             color = pheatmap.colors, 
             annotation_colors = ann.colors,
             breaks = breaksList, border_color = NA,
             gaps_col = cumsum(table(clusters.BEAT[new.order.BEAT])[unique(clusters.BEAT[new.order.BEAT])]))
pmap.BEAT
```

```{r include=FALSE}
if(saveFigs){
    pdf("output/Music_BEAT_deconvolution_via_integrated_pseudobulk_simple_clustering.pdf", width = 18, height = 6.5)
    pmap.BEAT
    dev.off()
}
```

```{r , fig.height=5}
temp.BEAT <- cbind(meta.BEAT, Blast.predicted = ann.row.BEAT$Blast.predicted)
temp.BEAT$status <- ifelse(temp.BEAT$event == "Alive", 0,ifelse(temp.BEAT$event == "Dead", 1, NA))

# # redefine groups <15 patients as 'Others'
# temp.BEAT <- temp.BEAT %>%  dplyr::add_count(`Primary Diagnosis`)
# temp.BEAT$cluster.refined <- ifelse(temp.BEAT$n > 15, temp.BEAT$cluster, "Others")

temp.BEAT$ELN <- factor(temp.BEAT$ELN, levels = c("Favorable", "Intermediate", "Adverse"))

km.beat.eln <- survfit(Surv(overal_survival, status) ~ ELN, data = temp.BEAT)
km.beat.eln

plot.beat.eln <- ggsurvplot(km.beat.eln,   
          pval = TRUE, # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("ELN=", "", names(km.beat.eln$strata)),
           
           palette = c("Favorable" = "#33CC00FF", "Intermediate" = "#CC9900FF", "Adverse" = "#991A00FF"), 
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

plot.beat.eln
```

```{r}
if(saveFigs){
  pdf("output/BEAT_ELN_KM.pdf")
  print(plot.beat.eln, newpage = F)
  dev.off()
}
```

```{r fig.height=5}
# For assigning samples with cut
qtiles[1] <- -Inf 
qtiles[4] <-  Inf 

X.test <- data.frame(deconv.BEAT[, c("GMP", "CD14 Mono", "HSC")], ELN = temp.BEAT$ELN, check.names = F)


preds.test <- predict(cox.tcga.deconv, newdata = X.test)

patient.groups <- cut(preds.test, qtiles, labels = c("Low Risk", "Medium Risk", "High Risk"))
temp.BEAT$patient.groups <- patient.groups

km.beat.qtile <- survfit(Surv(overal_survival, status) ~ patient.groups, data = temp.BEAT)
km.beat.qtile

plot.beat.ip.eln.prediction <- ggsurvplot(km.beat.qtile, 
           pval = T, 
           # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("patient.groups=", "", names(km.beat.qtile$strata)),
           
           palette = c("Low Risk" = "#33CC00FF", "Medium Risk" = "#CC9900FF", "High Risk" = "#991A00FF"), 
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))
plot.beat.ip.eln.prediction
```

```{r}
if(saveFigs){
  pdf("output/BEAT_IP_ELN.pdf")
  print(plot.beat.ip.eln.prediction, newpage = F)
  dev.off()
}
```

```{r alluvial BEAT}
library(scales)

plot.beat.alluvial <- 
temp.BEAT %>% dplyr::count(ELN, patient.groups, name = "Freq") %>% 
ggplot(aes(axis1 = ELN, axis2 = patient.groups, y = Freq)) +
  geom_flow(aes(fill = ELN)) + 
  geom_stratum() + 
  geom_text(stat = "stratum", aes(label = paste0(after_stat(stratum),"\n", percent(after_stat(prop), accuracy = 0.1))), size = 3) +
  theme_void() + scale_fill_manual(values = c("Favorable" = "#33CC00FF", "Intermediate" = "#CC9900FF", "Adverse" = "#991A00FF"))
plot.beat.alluvial
```

```{r BEAT Forest plot, fig.width=14, fig.height=10}
pdf("output/BEAT_Alluvial.pdf")
plot.beat.alluvial
dev.off()
```

### TARGET-AML

```{r TARGET deconvolution, cache=TRUE}
count.TARGET <- read.csv("data/Studies_Latest/TARGET/TARGET_ht_seq.csv")

count.TARGET <- wrangleMat(count.TARGET)

meta.TARGET <- read.csv("data/Studies_Latest/TARGET/TARGET_clinical_mut_meta.csv", row.names = 1)
meta.TARGET$sample <- rownames(meta.TARGET)

meta.TARGET$APS <- calculateAPS(count.TARGET)
meta.TARGET$Stemness <- calculateStemness(count.TARGET)

meta.TARGET$primary_diagnosis_refined <- meta.TARGET$primary_diagnosis

meta.TARGET[meta.TARGET$primary_diagnosis == "Mutated NPM1", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with mutated NPM1"
meta.TARGET[meta.TARGET$primary_diagnosis == "PML-RARa", "primary_diagnosis_refined"] <- 
    "Acute promyelocytic leukaemia, PML-RAR-alpha"
meta.TARGET[meta.TARGET$primary_diagnosis == "inv(16) CBF-beta/MYH11", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia, CBF-beta/MYH11"
meta.TARGET[meta.TARGET$primary_diagnosis == "t(8;21) RUNX1-RUNX1T1", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(8;21)(q22;q22); RUNX1-RUNX1T1"
meta.TARGET[meta.TARGET$primary_diagnosis == "t(9;11) MLLT3-MLL", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(9;11)(p22;q23); MLLT3-MLL"
meta.TARGET[meta.TARGET$primary_diagnosis == "Mutated CEBPA", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with mutated CEBPA"
meta.TARGET[meta.TARGET$primary_diagnosis == "t(6;9) DEK-NUP214", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(6;9)(p23;q34); DEK-NUP214"
meta.TARGET[meta.TARGET$primary_diagnosis == "t(3;5) NPM1-MLF1", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with myelodysplasia-related changes"
meta.TARGET[meta.TARGET$primary_diagnosis == "KMT2A-MLLT4", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(6;11)(q27;q23.3); MLLT4-KMT2A"
meta.TARGET[meta.TARGET$primary_diagnosis == "KMT2A-MLLT10", "primary_diagnosis_refined"] <- 
    "Acute myeloid leukemia with t(10;11)(p12;q23.3); MLLT10-KMT2A"


# mut.TARGET <- read.csv("data/Other studies/TARGET/TARGET_mutations_df_3.csv", row.names = 1)

meta.TARGET <- meta.TARGET[!grepl("Healthy", meta.TARGET$primary_diagnosis_refined ),]
count.TARGET.healthy <- count.TARGET[, setdiff(colnames(count.TARGET), rownames(meta.TARGET))]
count.TARGET <- count.TARGET[, rownames(meta.TARGET)]

meta.TARGET <- meta.TARGET %>% mutate(Age = Age.at.Diagnosis.in.Days/365, 
                                  Sex = ifelse(Gender == "Male", "male", "female"), 
                                  Recurrence = ifelse(First.Event == "Relapse", "Yes", 
                                                      ifelse(First.Event == "Censored", "No", "Induction Failure")),
                                  Blast = Bone.marrow.leukemic.blast.percentage....,
                                  Origin = "BM",
                                  FAB = FAB.Category,
                                  `Primary Diagnosis` = primary_diagnosis_refined, 
                                  overal_survival = Overall.Survival.Time.in.Days,
                                  ELN = Risk.group,
                                  Patient = TARGET.USI,
                                  Sample = sample,
                                  Other.Sample = NA,
                                  Study = "TARGET", 
                                  Status = ifelse(Vital.Status == "Dead",1,0))

meta.TARGET$ELN[meta.TARGET$ELN == "Unknown"] <- NA
meta.TARGET$ELN <- factor(meta.TARGET$ELN, labels = c("Adverse", "Favorable", "Intermediate"))
meta.TARGET$FAB[meta.TARGET$FAB %in% c("Unknown", "NOS") ] <- NA


# primary diagnoses
pdiagnoses.TARGET <- stats::model.matrix(~ 0 + meta.TARGET$primary_diagnosis_refined)
colnames(pdiagnoses.TARGET) <- sub(pattern = "meta.TARGET\\$primary_diagnosis_refined",
                                 replacement = "", colnames(pdiagnoses.TARGET))




# ens to symbol map
ens2gene <- cinaR::grch38
m <- match(rownames(count.TARGET), ens2gene$ensgene)
mapped.genes <- ens2gene$symbol[m]

# duplicated name/NA/mitochondrial genes
removed.genes <- duplicated(mapped.genes) | is.na(mapped.genes) | grepl("^MT", mapped.genes)

count.TARGET <- count.TARGET[!removed.genes,]
rownames(count.TARGET) <- mapped.genes[!removed.genes]


T.eset <- ExpressionSet(assayData = as.matrix(count.TARGET))

deconv.TARGET <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample', 
                                            verbose = F)$Est.prop.weighted
```

```{r fig.width=15, fig.height=5}
ann.row.TARGET <- as.data.frame(cbind.data.frame(pdiagnoses.TARGET, 
                                      Blast = meta.TARGET$Bone.marrow.leukemic.blast.percentage...., 
                                      Stemness = meta.TARGET$Stemness, 
                                      FAB = meta.TARGET$FAB,
                                      ELN = meta.TARGET$ELN))

rownames(ann.row.TARGET) <- colnames(count.TARGET)


# assignment threshold
clusters.TARGET <- apply(deconv.TARGET, 1, function(x){names(x)[which(x == max(x))]})



# assigning colors for reproducibility 
ELN.colors = c("Favorable" = "#00b533", "Intermediate"= "#d6d300", "Adverse" = "#D5232F")
FAB.colors = paletteer_d("awtools::a_palette")
names(FAB.colors) <- paste0("M", 0:7)

Status.colors <- DiscretePalette(4)
names(Status.colors) <- unique(ann.row.TARGET$Status)

ann.colors <- list(cluster = cell.type.colors[ unique(clusters.TARGET)], 
                   ELN = ELN.colors, FAB = FAB.colors, Status = Status.colors)



# new.order.TARGET <- names(clusters.TARGET %>% sort)
new.order.TARGET <- patientOrder(deconv.TARGET, clusters.TARGET, cell.type.order)


breaksList = seq(0,1, by = .001)
pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))

pmap.TARGET <- pheatmap(t(deconv.TARGET [new.order.TARGET, intersect(cell.type.order, colnames(deconv.TARGET))]),
         cluster_cols = F, cluster_rows = F, show_rownames = T, show_colnames = F,
         annotation_col = cbind(ann.row.TARGET[new.order.TARGET,], cluster = clusters.TARGET[new.order.TARGET]),
         annotation_colors = ann.colors,
         color = pheatmap.colors, breaks = breaksList, border_color = NA, 
         gaps_col = cumsum(table(clusters.TARGET[new.order.TARGET])[unique(clusters.TARGET[new.order.TARGET])]))

pmap.TARGET
```

```{r include=FALSE}
if(saveFigs){
    pdf("output/Music_TARGET_deconvolution_via_integrated_pseudobulk_simple_clustering.pdf", height = 7, width = 16)
    print(pmap.TARGET, newpage = F)
    dev.off()
}
```

```{r , fig.height=5}
temp.TARGET <- meta.TARGET
temp.TARGET$status <- ifelse(temp.TARGET$Vital.Status == "Alive", 0,ifelse(temp.TARGET$Vital.Status == "Dead", 1, NA))

# # redefine groups <15 patients as 'Others'
# temp.TARGET <- temp.TARGET %>%  dplyr::add_count(`Primary Diagnosis`)
# temp.TARGET$cluster.refined <- ifelse(temp.TARGET$n > 15, temp.TARGET$cluster, "Others")

temp.TARGET$ELN <- factor(temp.TARGET$ELN, levels = c("Favorable", "Intermediate", "Adverse"))

km.target.eln <- survfit(Surv(overal_survival, status) ~ ELN, data = temp.TARGET)
km.target.eln

plot.target.eln <- ggsurvplot(km.target.eln,   
          pval = TRUE, # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("ELN=", "", names(km.target.eln$strata)),
           
           palette = c("Favorable" = "#33CC00FF", "Intermediate" = "#CC9900FF", "Adverse" = "#991A00FF"), 
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))
plot.target.eln
```

```{r}
if(saveFigs){
  pdf("output/TARGET_ELN.pdf")
  print(plot.target.eln, newpage = F)
  dev.off()
}
```

```{r , fig.height=5}

X.test.target <- data.frame(deconv.TARGET[, c("GMP", "CD14 Mono", "HSC")], ELN = temp.TARGET$ELN, check.names = F)


preds.test <- predict(cox.tcga.deconv, newdata = X.test.target)

patient.groups <- cut(preds.test, qtiles, labels = c("Low Risk", "Medium Risk", "High Risk"))
temp.TARGET$patient.groups <- patient.groups

km.target.qtile <- survfit(Surv(overal_survival, status) ~ patient.groups, data = temp.TARGET)
km.target.qtile

plot.target.ip.eln.prediction <- ggsurvplot(km.target.qtile, 
           pval = T, 
           # Change legends: title & labels
          legend.title = "Cluster",
          legend.labs = gsub("patient.groups=", "", names(km.target.qtile$strata)),
           
           palette = c("Low Risk" = "#33CC00FF", "Medium Risk" = "#CC9900FF", "High Risk" = "#991A00FF"), 
           
           # # Add risk table
           risk.table = TRUE,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))
plot.target.ip.eln.prediction
```

```{r}
if(saveFigs){
  pdf("output/TARGET_IP_ELN.pdf")
  print(plot.target.ip.eln.prediction, newpage = F)
  dev.off()
}
```

```{r alluvial TARGET}


plot.target.alluvial <- 
temp.TARGET %>% dplyr::count(ELN, patient.groups, name = "Freq") %>% 
ggplot(aes(axis1 = ELN, axis2 = patient.groups, y = Freq)) +
  geom_flow(aes(fill = ELN)) + 
  geom_stratum() + 
  geom_text(stat = "stratum", aes(label = paste0(after_stat(stratum),"\n", percent(after_stat(prop), accuracy = 0.1))), size = 3) +
  theme_void() + scale_fill_manual(values = c("Favorable" = "#33CC00FF", "Intermediate" = "#CC9900FF", "Adverse" = "#991A00FF"))
plot.target.alluvial
```

```{r TARGET Alluvial plot, fig.width=14, fig.height=10}
pdf("output/TARGET_Alluvial.pdf")
plot.target.alluvial
dev.off()
```

## LEUCEGENE

```{r}
count.LEUCE <- read.csv("data/Other studies/LEUCEGENE/Leucegene_count_matrix.csv", check.names = F)

count.LEUCE <- wrangleMat(count.LEUCE)


meta.LEUCE <- read.csv("data/Other studies/LEUCEGENE/Leucegene_meta_data.csv", row.names = 1)
meta.LEUCE <- read.csv("data/Other studies/LEUCEGENE/Leucegene_Annotations.tsv", sep = "\t")

rownames(meta.LEUCE) <- meta.LEUCE$sample_id

meta.LEUCE <- meta.LEUCE[colnames(count.LEUCE), ]

# Not all genes exist in LEUCEGENE for APS and Stemness
meta.LEUCE$APS <- NA
meta.LEUCE$Stemness <- NA

meta.LEUCE$FAB <- ifelse(meta.LEUCE$FAB.classification == "Not classifiable by FAB criteria", NA,meta.LEUCE$FAB.classification)
meta.LEUCE$FAB <- sub(".*-", "", meta.LEUCE$FAB)  %>% substr(1,2) 

meta.LEUCE$primary_diagnosis <- meta.LEUCE$WHO.classification

meta.LEUCE$primary_diagnosis <- ifelse(meta.LEUCE$primary_diagnosis %in% c("Acute myeloid leukemia, NOS",
                                                                   "\"\"Acute myeloid leukaemia, NOS\"\"",
                                                                   "Acute myeloid leukaemia, NOS",
                                                                   "AML with maturation",
                                                                   "Acute myeloid leukemia with maturation",
                                                                   "AML without maturation",
                                                                   "Acute myeloid leukemia without maturation",
                                                                   "AML with minimal differentiation",
                                                                   "Acute myeloid leukemia, minimal differentiation",
                                                                   "Acute myelomonocytic leukaemia",
                                                                   "Acute myelomonocytic leukemia",
                                                                   "Acute monoblastic and monocytic leukemia",
                                                                   "Acute monoblastic and monocytic leukaemia",
                                                                   "Acute erythroid leukaemia",
                                                                   "Acute megakaryoblastic leukaemia",
                                                                   "Acute basophilic leukemia",
                                                                   "Acute panmyelosis with myelofibrosis"), 
                                              "Acute myeloid leukemia, NOS", meta.LEUCE$primary_diagnosis)


meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("AML with inv(16)(p13.1q22) or t(16;16)(p13.1;q22); CBFB-MYH11"), "primary_diagnosis"] <- 
    "Acute myeloid leukemia, CBF-beta/MYH11"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("AML with inv(3)(q21q26.2) or t(3;3)(q21;q26.2); RPN1-EVI1"), "primary_diagnosis"] <- 
    "Acute myeloid leukemia with inv(3)(q21q26.2) or t(3;3)(q21;q26.2); RPN1-EVI1"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("AML with t(8;21)(q22;q22); RUNX1-RUNX1T1"), "primary_diagnosis"] <- 
    "Acute myeloid leukemia with t(8;21)(q22;q22); RUNX1-RUNX1T1"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("AML with t(9;11)(p22;q23); MLLT3-MLL"), "primary_diagnosis"] <- 
    "Acute myeloid leukemia with t(9;11)(p22;q23); MLLT3-MLL"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("Acute promyelocytic leukaemia with t(15;17)(q24;q21); PML-RARA"), "primary_diagnosis"] <- 
    "Acute promyelocytic leukaemia, PML-RAR-alpha"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("AML with t(6;9)(p23;q34); DEK-NUP214"), "primary_diagnosis"] <- 
    "Acute myeloid leukemia with t(6;9)(p23;q34); DEK-NUP214"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in% 
             c("AML with myelodysplasia-related changes"), "primary_diagnosis"] <-
    "Acute myeloid leukemia with myelodysplasia-related changes"

meta.LEUCE[meta.LEUCE$primary_diagnosis %in%
             c("Therapy-related myeloid neoplasms"), "primary_diagnosis"] <-
    "Therapy related myeloid neoplasm"

meta.LEUCE <- meta.LEUCE %>% mutate(Age = Age.at.sampling, 
                                  Sex = ifelse(Sex == "F", "female", "male"), 
                                  Recurrence = NA,
                                  Blast = Blasts....,
                                  Origin = ifelse(Tissue == "Bone marrow", "BM", "PB"),
                                  `Primary Diagnosis` = primary_diagnosis, 
                                  ELN = NA, 
                                  Status = NA,
                                  overal_survival = NA,
                                  Patient = sample_id,
                                  Sample = sample_id,
                                  Other.Sample = BioSample_ID,
                                  Study = "LEUCE") 


# primary diagnoses
pdiagnoses.LEUCE <- stats::model.matrix(~ 0 + meta.LEUCE$primary_diagnosis)
colnames(pdiagnoses.LEUCE) <- sub(pattern = "meta.LEUCE\\$primary_diagnosis", 
                                 replacement = "", colnames(pdiagnoses.LEUCE))



T.eset <- ExpressionSet(assayData = as.matrix(count.LEUCE))

deconv.LEUCE <- music_prop(bulk.eset = T.eset, sc.eset = C.eset, clusters = 'label.new',
                                            markers = NULL, normalize = FALSE, samples = 'Sample', 
                                            verbose = F)$Est.prop.weighted
```

```{r}


# LEUCEGENE patient annotations 
ann.row.LEUCE <- as.data.frame(cbind.data.frame(pdiagnoses.LEUCE, 
                                    FAB = meta.LEUCE[,"FAB", drop = F]
                                    ))

rownames(ann.row.LEUCE) <- colnames(count.LEUCE)


# assignment threshold
# clusters.LEUCE <- apply(deconv.LEUCE > 0.5, 1, function(x){
#   ifelse(length(x[x]) > 0, names(x[x]), "Mixed Profiles")
# }) 

clusters.LEUCE <- apply(deconv.LEUCE, 1, function(x){names(x)[which(x == max(x))]})


FAB.colors = paletteer_d("awtools::a_palette")
names(FAB.colors) <- paste0("M", 0:7)
ann.colors <- list(cluster = cell.type.colors[ unique(clusters.LEUCE)], 
                   FAB = FAB.colors)


#new.order.LEUCE <- names(clusters.LEUCE %>% sort)
new.order.LEUCE <- patientOrder(deconv.LEUCE, clusters.LEUCE, cell.type.order)


breaksList = seq(0,1, by = .001)
pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))
pmap.LEUCE <- pheatmap(t(deconv.LEUCE [new.order.LEUCE, intersect(cell.type.order, colnames(deconv.LEUCE))]),
         cluster_cols = F, cluster_rows = F, show_rownames = T, show_colnames = F,
         annotation_col = cbind(ann.row.LEUCE[new.order.LEUCE,], cluster = clusters.LEUCE[new.order.LEUCE]),
         annotation_colors = ann.colors,
         color = pheatmap.colors, breaks = breaksList, border_color = NA, 
         gaps_col = cumsum(table(clusters.LEUCE[new.order.LEUCE])[unique(clusters.LEUCE[new.order.LEUCE])]))

pmap.LEUCE
```

```{r include=FALSE}
if(saveFigs){
    pdf("output/Music_LEUCEGENE_deconvolution_via_integrated_pseudobulk_simple_clustering.pdf", height = 5, width = 17)
    print(pmap.LEUCE, newpage = F)
    dev.off()
}
```

```{r}

# EDUCATED GUESS: Deconvolution correlates with FAB therefore no survival differences
# So, let's focus on mutation combinations that explains the cell types.
library(data.table)
mut.TCGA <- fread("data/Studies_Latest/TCGA/TCGA_mutations.txt")
mut.TCGA <- mut.TCGA %>% mutate(Patient = Matched_Norm_Sample_Barcode)
mut.TCGA <- reshape2::acast(mut.TCGA, Patient ~ Hugo_Symbol)
meta.TCGA$Order <- 1:nrow(meta.TCGA)
mut.TCGA <- merge(meta.TCGA[, c("Patient", "Sample", "Order"), drop = F], mut.TCGA, by.x = "Patient", by.y = 0,  all.x = T)
mut.TCGA <- mut.TCGA[order(mut.TCGA$Order),]
rownames(mut.TCGA) <- mut.TCGA$Sample
mut.TCGA <- mut.TCGA[,-c(1:3)]

mut.BEAT <- fread("data/Studies_Latest/BEAT-AML/BEAT_mutations.txt")
mut.BEAT$Patient <- strsplit(mut.BEAT$Tumor_Sample_Barcode, "_", fixed = T) %>% sapply(function(x) x[length(x)])
mut.BEAT <- reshape2::acast(mut.BEAT, Patient ~ Hugo_Symbol)
meta.BEAT$Order <- 1:nrow(meta.BEAT)
mut.BEAT <- merge(meta.BEAT[,c("Patient", "Sample", "LLS_SampleID", "Order" ), drop = F], mut.BEAT, 
                  by.x = "LLS_SampleID", by.y = 0,  all.x = T)[,-1]
mut.BEAT <- mut.BEAT[order(mut.BEAT$Order),]
rownames(mut.BEAT) <- mut.BEAT$Sample
mut.BEAT <- mut.BEAT[,-c(1:3)]


mut.TARGET <- fread("data/Studies_Latest/TARGET/TARGET_mutations.txt")
mut.TARGET$Patient <- strsplit(mut.TARGET$Tumor_Sample_Barcode, "-", fixed = T) %>% 
  sapply(function(x) paste0(x[-length(x)], collapse = "-"))
mut.TARGET <- reshape2::acast(mut.TARGET, Patient ~ Hugo_Symbol)
# keep order
meta.TARGET$Order <- 1:nrow(meta.TARGET)
mut.TARGET <- merge(meta.TARGET[,c("Patient", "Sample", "Order"), drop = F], mut.TARGET, by.x = "Patient", by.y = 0,  all.x = T)
mut.TARGET <- mut.TARGET[order(mut.TARGET$Order),]
rownames(mut.TARGET) <- mut.TARGET$Sample
mut.TARGET <- mut.TARGET[,-c(1:3)]


# Mutations for LUMC
mut.LUMC <- as.data.frame(fread("data/Studies_Latest/mutations_oh.csv", drop = 1))
mut.LUMC <- mut.LUMC %>% filter(grepl("AML.",sample_ID))

mut.LUMC$NRAS <- mut.LUMC$NRAS_all
mut.LUMC$IDH2 <- mut.LUMC$IDH2_all
mut.LUMC$ASXL1 <- mut.LUMC$ASXL1_all
mut.LUMC$DNMT3A <- mut.LUMC$DNMT3A_all
mut.LUMC$IDH1 <- mut.LUMC$IDH1.R132 + mut.LUMC$IDH1.R172
mut.LUMC$FLT3 <- mut.LUMC$FLT3.TKD + mut.LUMC$FLT3.ITD + mut.LUMC$FLT3
mut.LUMC$KIT <- mut.LUMC$KIT_all

mut.LUMC <- as.data.frame(mut.LUMC)

rownames(mut.LUMC) <- mut.LUMC$sample_ID
mut.LUMC <- mut.LUMC[,-1]


# Mutations for LEUCE
mut.LEUCE <- meta.LEUCE[,c("IDH1.R132.mutation", "NPM1.mutation", "FLT3.ITD.km.curated")]
colnames(mut.LEUCE) <- c("IDH1", "NPM1", "FLT3")
# mut.LEUCE <- mut.LEUCE %>% filter(grepl("^X",sample_ID))

mut.train <- plyr::rbind.fill(
  mut.TCGA,
  mut.LUMC,
  mut.BEAT,
  mut.TARGET,
  mut.LEUCE
)

rownames(mut.train) <- c(rownames(mut.TCGA), 
                         rownames(mut.LUMC), 
                         rownames(mut.BEAT),
                         rownames(mut.TARGET),
                         rownames(mut.LEUCE))

selected.mutations <- colSds(as.matrix(mut.train), na.rm = T) > 0
selected.mutations[is.na(selected.mutations)] <- FALSE

mut.train <- mut.train[, selected.mutations & (colSums(mut.train, na.rm = T) > 10)]

mut.train <- mut.train[,!grepl(".", colnames(mut.train), fixed = T) & !grepl("_all", colnames(mut.train), fixed = T)]

common.mutations <- colnames(mut.train)
```

### MERGE ALL STUDIES

```{r}
# merge patient profiles from TCGA and BEAT
common.features <- c("Sample", "Other.Sample", "Patient","Study","Age", "Blast", "Sex","Origin","Primary Diagnosis", 
                     "ELN", "FAB","Status", "overal_survival", "Stemness", "APS", "Recurrence")

meta.train <- plyr::rbind.fill(
  cbind(meta.TCGA[,common.features],cluster = clusters.TCGA),
  cbind(meta.LUMC[,common.features],cluster = clusters.LUMC),
  cbind(meta.BEAT[,common.features],cluster = clusters.BEAT),
  cbind(meta.TARGET[,common.features],cluster = clusters.TARGET),
  cbind(meta.LEUCE[,common.features],cluster = clusters.LEUCE)
)
# merge mutation data with meta
meta.train <- cbind(meta.train, mut.train)
rownames(meta.train) <- NULL

pdiagnoses.train <- stats::model.matrix(~ 0 + meta.train$`Primary Diagnosis`)
colnames(pdiagnoses.train) <- sub(pattern = "meta.train\\$`Primary Diagnosis`", 
                                 replacement = "", colnames(pdiagnoses.train))


rownames(meta.train) <- gsub("-", ".",meta.train$Sample)

deconv.train <- rbind(
  deconv.TCGA,
  deconv.LUMC,
  deconv.BEAT,
  deconv.TARGET,
  deconv.LEUCE
)


# new.order.TARGET <- names(clusters.TARGET %>% sort)
new.order.train <- patientOrder(deconv.train, meta.train$cluster, cell.type.order)


breaksList = seq(0,1, by = .001)
pheatmap.colors <- colorRampPalette((brewer.pal(n = 7, name = "Blues")))(length(breaksList))

age.color <- colorRampPalette((brewer.pal(n = 7, name = "Reds")))(length(seq(0,100, by = 1)))


train.clusters <- meta.train[new.order.train,"cluster"]

study.colors <- as.vector(paletteer_d(`"ggsci::default_uchicago"`))
names(study.colors) <- unique(meta.train$Study)
study.colors <- study.colors[!is.na(names(study.colors))]

ann.colors <- list(cluster = cell.type.colors[ unique(meta.train$cluster)], 
                   ELN = ELN.colors, FAB = FAB.colors, Status = Status.colors, 
                   Origin = c("BM"="#A3B0E4", "PB"="#F0535A"), Age = age.color, 
                   Sex = c("male" = "#72CAF0", "female" =  "#EE846D"), 
                   Study = study.colors)
ann.row.train <- cbind.data.frame(pdiagnoses.train, 
                                  meta.train[, c("Study","Origin","Age", "Sex","FAB", "ELN",  "cluster")])

rownames(ann.row.train) <- rownames(meta.train)

pmap.train <- pheatmap(t(deconv.train [new.order.train, intersect(cell.type.order, colnames(deconv.train))]),
                       cluster_cols = F, cluster_rows = F, show_rownames = T, show_colnames = F,
                       annotation_col = ann.row.train[new.order.train,],
                       annotation_colors = ann.colors,
                       color = pheatmap.colors, breaks = breaksList, border_color = NA,
                       gaps_col = cumsum(table(train.clusters)[unique(train.clusters)])
         
         )

pmap.train
```

```{r}
if(saveFigs){
  
  pdf("output/ALL_merged_deconv.pdf", width = 20, height = 7)
  pmap.train
  dev.off()
}
```

```{r}
library(gtsummary)
summary.table.train <- meta.train %>% 
  filter(!duplicated(Patient)) %>% 
  dplyr::select(Age, Sex, Study, FAB,ELN, `Primary Diagnosis` ) %>% 
  
  gtsummary::tbl_summary(by = "Study" ,
                         statistic = list(all_continuous() ~ "{median} ({min}, {max})",
                                          all_categorical() ~ "{n}/{N} ({p}%)"))  %>% 
  add_overall() %>%   
  bold_labels() %>% 
  modify_footnote(
    all_stat_cols() ~ "Median (Range) or Frequency (%)"
  ) 
  
summary.table.train %>% as_gt() %>% 
  gt::gtsave(filename = "output/Patient_Characteristics.pdf")
```

### DECONVOLUTION ON REDUCED DIMENTIONS

```{r}

noise <- matrix(runif(prod(dim(deconv.train)), min = -1e-6, max = 1e-6), nrow = nrow(deconv.train))
new.vals <- deconv.train + noise
tsne <- Rtsne::Rtsne(new.vals)

set.seed(2)

umap <- umap::umap(deconv.train)


plot.df <- data.frame(UMAP = umap$layout, TSNE = tsne$Y, deconv.train, meta.train, check.names = F)
# plot.df$`Primary Diagnosis` <- gsub("Acute myeloid leukemia", "", plot.df$`Primary Diagnosis`)

fontSize <- 12

pdf("output/UMAP_ALL_patients_Assignment.pdf", useDingbats = F)

plot.celltype <- ggplot(plot.df[sample(1:nrow(plot.df)),], aes(UMAP.1, UMAP.2, color = cluster)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + labs(color = "Assignment") +
  scale_color_manual(values = cell.type.colors[!names(cell.type.colors) %in% c("Others")])
plot.celltype
dev.off()

pdf("output/UMAP_ALL_patients_Sex.pdf")
plot.sex <- ggplot(plot.df[sample(1:nrow(plot.df)),], aes(UMAP.1, UMAP.2, color = Sex)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + 
  scale_color_manual(values = ann.colors$Sex)
plot.sex
dev.off()

pdf("output/UMAP_ALL_patients_Stemness.pdf")
plot.stemness <- ggplot(plot.df[sample(1:nrow(plot.df)),], aes(UMAP.1, UMAP.2, color = Stemness)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + 
  scale_color_viridis_c(na.value = NA)
plot.stemness
dev.off()

pdf("output/UMAP_ALL_patients_APS.pdf")
plot.APS <- ggplot(plot.df[sample(1:nrow(plot.df)),], aes(UMAP.1, UMAP.2, color = APS)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + 
  scale_color_viridis_c(option = "B", na.value = NA)
plot.APS
dev.off()

pdf("output/UMAP_ALL_patients_Blast.pdf")
plot.Blast <- ggplot(plot.df[sample(1:nrow(plot.df)),], aes(UMAP.1, UMAP.2, color = Blast)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + 
  scale_color_viridis_c(option = "D", na.value = NA)
plot.Blast
dev.off()

pdf("output/UMAP_ALL_patients_Study.pdf")
plot.study <- ggplot(plot.df[sample(1:nrow(plot.df)),] , aes(UMAP.1, UMAP.2, color = Study)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + scale_color_manual(values = ann.colors$Study)
plot.study
dev.off()

pdf("output/UMAP_ALL_patients_Age.pdf")
plot.age<- ggplot(plot.df[sample(1:nrow(plot.df)),] , aes(UMAP.1, UMAP.2, color = Age)) +
  geom_point(size = 2) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + scale_color_gradient(low = "lightblue", high = "darkblue")
plot.age
dev.off()

pdf("output/UMAP_ALL_patients_FAB.pdf")
plot.fab <- ggplot(plot.df[sample(1:nrow(plot.df)),] , aes(UMAP.1, UMAP.2)) +
  geom_point(size = 2, data = plot.df[is.na(plot.df$FAB),], color = "gray") + 
  geom_point(size = 2, data = plot.df[!is.na(plot.df$FAB),], aes(color = FAB)) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + 
  scale_color_manual(values = ann.colors$FAB)
plot.fab
dev.off()

pdf("output/UMAP_ALL_patients_ELN.pdf")
plot.eln<- ggplot(plot.df[sample(1:nrow(plot.df)),] , aes(UMAP.1, UMAP.2)) +
  geom_point(size = 2, data = plot.df[is.na(plot.df$ELN),], color = "gray") + 
  geom_point(size = 2, data = plot.df[!is.na(plot.df$ELN),], aes(color = ELN)) +
  theme_light(base_size = fontSize) + 
  coord_fixed(ratio = 1) + scale_color_manual(values = ann.colors$ELN)
plot.eln
dev.off()

pdf("output/UMAP_ALL_patients_PrimaryDiagnosis.pdf")
plot.primary<- ggplot(plot.df[sample(1:nrow(plot.df)),] , aes(UMAP.1, UMAP.2)) +
  geom_point(data=dplyr::select(plot.df[sample(1:nrow(plot.df)),], -`Primary Diagnosis`), color = "gray") + 
  geom_point(size=1, pch=21, aes(fill=`Primary Diagnosis`)) +
  ggforce::facet_wrap_paginate(~`Primary Diagnosis`, page = 14) + 
  scale_fill_manual(values = DiscretePalette(20)) + 
  theme.gg + theme(legend.position = "none")
plot.primary
dev.off


lapply(cell.type.order, function(x){
  p1 <- ggplot(plot.df[sample(1:nrow(plot.df)),] , aes(UMAP.1, UMAP.2, color = .data[[x]])) +
  geom_point(size = 2) +
  theme_bw(base_size = fontSize) +
  coord_fixed(ratio = 1) + 
  ggtitle(x) +  theme(legend.position="bottom") + 
  scale_color_viridis_c(option = "C", na.value = NA, breaks = c(0,1)) 
  ggsave(paste0("output/UMAP_ALL_patients_celltypes_", x,".pdf"), plot = p1, width = 6, height = 6)
})

```

```{r Pileup FAB}
df.stacked.violin <- plot.df[, c("Sample","FAB", cell.type.order)] %>% 
  reshape2::melt(id.vars = c("Sample", "FAB"), measure.vars = cell.type.order,
                       variable.name = "Cell Type", value.name = "Perc")
library(cowplot)


pdf("output/FAB_vs_CellType_Stacked_Violin.pdf")
ggplot(df.stacked.violin %>% filter(!is.na(FAB)), aes(`Cell Type`, Perc, fill = FAB)) +
        geom_violin(scale = "width", adjust = 1, trim = TRUE) +
        scale_y_continuous(expand = c(0, 0), position="right", labels = function(x)
                           c(rep(x = "", times = length(x)-2), x[length(x) - 1], "")) +
        facet_grid(rows = vars(FAB), switch = "y") +
        theme_cowplot(font_size = 12) +
        theme(legend.position = "none", panel.spacing = unit(0, "lines"),
              panel.background = element_rect(fill = NA, color = "black"),
              strip.background = element_blank(),
              strip.text = element_text(face = "bold"),
              strip.text.y.left = element_text(angle = 0),
              axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + 
  xlab("") + ylab("Cell Type %") + 
  scale_fill_manual(values = ann.colors$FAB)
  dev.off()
```

```{r}
quantile(plot.df$Stemness, na.rm = T, c(0, 0.33, 0.66, 1))
plot.df$Stemness.categorical <- cut(plot.df$Stemness, quantile(plot.df$Stemness, na.rm = T, c(0, 0.5, 1)), labels = c("Low", "High"),
    include.lowest = T)

df.stacked.stemness <- plot.df[, c("Sample","Stemness.categorical","Study", cell.type.order)] %>% 
  reshape2::melt(id.vars = c("Sample", "Stemness.categorical", "Study"), measure.vars = cell.type.order,
                       variable.name = "Cell Type", value.name = "Perc")


pdf("output/Stemness_vs_CellType_Stacked_Violin.pdf", width = 12)
ggplot(df.stacked.stemness %>% filter(!is.na(Stemness.categorical)), aes(`Cell Type`, Perc, fill = `Cell Type`)) +
        geom_violin(scale = "width", adjust = 1, trim = TRUE) +
        scale_y_continuous(expand = c(0, 0), position="right", labels = function(x)
                           c(rep(x = "", times = length(x)-2), x[length(x) - 1], "")) +
        facet_grid(rows = vars(Stemness.categorical), switch = "y", cols = vars(Study)) +
        theme_cowplot(font_size = 12) +
        theme(legend.position = "none", panel.spacing = unit(0, "lines"),
              panel.background = element_rect(fill = NA, color = "black"),
              strip.background = element_blank(),
              strip.text = element_text(face = "bold"),
              strip.text.y.left = element_text(angle = 0),
              axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5)) + 
  xlab("") + ylab("Cell Type %") + 
  scale_fill_manual(values = ann.colors$cluster)
  dev.off()
  

  
  
lm.cell.types <-df.stacked.stemness %>% filter(Study != "LEUCE") %>% 
  split(list(.$Study, .$`Cell Type`)) %>% 
  purrr::map(~ lm(Perc ~ Stemness.categorical, data = .))
  
df.lm.cell.types <- lm.cell.types %>% 
  purrr::map(summary) %>% 
  purrr::map_df(broom::tidy) %>% 
  dplyr::filter(term != "(Intercept)")
  
df.lm.cell.types$Study    <- strsplit(names(lm.cell.types), ".", fixed = T) %>% purrr::map(1) %>% do.call(c, .)
df.lm.cell.types$CellType <- strsplit(names(lm.cell.types), ".", fixed = T) %>% purrr::map(2) %>% do.call(c, .)

pdf("output/Cell_type_vs_Stemness_Slope.pdf")

ggplot(df.lm.cell.types, aes(Study, CellType, fill = estimate,label = ifelse(p.value < 0.05, "*", ""))) + geom_tile() + geom_text(size = 5, color = "white", nudge_y = -0.1) + theme_bw(base_size = 10) + 
  scale_fill_gradient2(low = "#085ea8", mid = "white", high="#d60e00") + 
  labs(fill = "Slope", caption = "* p < 0.05") + xlab("") + ylab("")

ggplot(df.lm.cell.types, aes(Study, CellType, fill = ifelse(estimate > 0, "Positive", "Negative"),label = ifelse(p.value < 0.05, "*", ""))) + geom_tile() + geom_text(size = 5, color = "white", nudge_y = -0.1) + theme_bw(base_size = 10) + 
  scale_fill_manual(values = c("#085ea8", "#d60e00")) + 
  labs(fill = "Slope", caption = "* p < 0.05") + xlab("") + ylab("")
dev.off()

lm(Perc ~ Stemness.categorical, data = df.stacked.stemness)
```

```{r UMAP mutations}
plot.df[intersected.mutations] <- lapply(plot.df[intersected.mutations] , 
                                         function(x){factor(ifelse(x > 0, "Yes", "No"), levels = c("Yes", "No"))})

pdf("output/UMAP_Common_Mutations.pdf", width = 10, height = 5)
lapply(c("NPM1", "FLT3", "IDH2", "WT1", "TET2", "JAK2", "MYC", "NRAS"), function(x){
  
  p1 <- ggplot(plot.df[rev(order(plot.df[,x])),] , aes(UMAP.1, UMAP.2)) +
  geom_point(size=1, data=dplyr::select(plot.df, -.data[[x]]), color = "gray") +
  geom_point(size=2, aes(color=.data[[x]])) + 
  scale_color_manual(values = c("Red", "Gray"), na.value = NA) + 
  theme.gg + ggtitle(x) 
  
  p2 <- ggplot(plot.df[rev(order(plot.df[,x])),] %>% filter(.data[[x]] == "Yes"), aes(UMAP.1, UMAP.2)) +
  stat_density_2d(geom = "polygon", contour = TRUE,
                  aes(fill = after_stat(level)), colour = "black",
                  bins = 5)+
  geom_point(size =2, color = "Red")+
  scale_fill_distiller(palette = "Blues", direction = 1) +
  theme.gg +  ggtitle(x) 
  
  ggarrange(p1, p2)
  
})
dev.off()
```

```{r Mutation vs Cell Type changes}


plot.df[c(cell.type.order,"NPM1")] %>% filter_all(~ !is.na(.)) %>% melt() %>% 
ggplot(aes(variable, value, fill = NPM1)) + geom_boxplot(outlier.shape = 1) + theme.gg + 
  stat_compare_means(aes(label = ..p.signif..),method = "wilcox.test", hide.ns = T) + 
  scale_x_discrete(guide = guide_axis(angle = 90))
```

### DRUG RESPONSE GRAPHS

```{r All spesific drug responses, message=FALSE}
meta.BEAT.drug <- readxl::read_excel("data/Studies_Latest/BEAT-AML/BEAT_Paper_meta.xlsx", sheet = "Table S10-Drug Responses")
meta.BEAT.drug <- split(meta.BEAT.drug %>% dplyr::select(-inhibitor), meta.BEAT.drug$inhibitor)

pdf("output/UMAP_Drug_response_AUC.pdf")
lapply(names(meta.BEAT.drug), function(drug.name){
  
  drug.df <- meta.BEAT.drug[[drug.name]]
  temp.plot <- merge(plot.df[,c("Sample","Other.Sample", "UMAP.1", "UMAP.2", "cluster", "NPM1")] , 
                     drug.df, by.x = "Other.Sample", by.y = "lab_id", all.x = T)
  
  
  plot.drug<- ggplot(temp.plot[sample(1:nrow(temp.plot)),] , aes(UMAP.1, UMAP.2, color = auc)) +
    geom_point(size = 2, data = temp.plot[ is.na(temp.plot$auc),], color = "gray") + 
    geom_point(size = 2, data = temp.plot[!is.na(temp.plot$auc),]) +
    theme.gg + ggtitle(drug.name) +
    coord_fixed(ratio = 1) + labs(color = "AUC") +
    scale_color_viridis_c(option = "G", direction = -1)
    
  temp.plot <- temp.plot %>% mutate(cluster.refined = ifelse(cluster %in% c("CD14 Mono", "HSC", "EMP", "T Cells", "GMP"), cluster, "Others"))
  
  plot.drug.cluster<- ggplot(temp.plot, aes(cluster.refined, auc, fill = cluster.refined)) +
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(size = 1, width = 0.1) +
    coord_flip() + 
    theme.gg + ggtitle(drug.name) + 
    scale_fill_manual(values = cell.type.colors) + 
    theme(legend.position = "none") + xlab("")
  
  p.val <- summary(aov(auc ~ cluster.refined, temp.plot))[[1]]$`Pr(>F)`[1]
  
  plot.drug | plot.drug.cluster + labs(caption = sprintf('anova p-value: %.5f', p.val))
  
  
}) 

dev.off()



# anova one way pvals
owa.p <- lapply(names(meta.BEAT.drug), function(drug.name){
  
  drug.df <- meta.BEAT.drug[[drug.name]]
  temp.plot <- merge(plot.df[,c("Sample","Other.Sample", "UMAP.1", "UMAP.2", "cluster", "NPM1")] , 
                     drug.df, by.x = "Other.Sample", by.y = "lab_id", all.x = T)
  
    
  temp.plot <- temp.plot %>% mutate(cluster.refined = ifelse(cluster %in% c("CD14 Mono", "HSC", "EMP", "T Cells", "GMP"), cluster, "Others"))
  
  p = summary(aov(auc ~ cluster.refined, temp.plot))[[1]]$`Pr(>F)`[1]
  return(cbind(Drug = drug.name, p = sprintf("%.2e", p ), logp = -log(p)))
  
}) %>% do.call(rbind, .) %>% as.data.frame

owa.p$logp <- as.numeric(owa.p$logp)
owa.p<- owa.p[order(-owa.p$logp),]
owa.p$adj.p <- p.adjust(owa.p$p, method = "BH")

writexl::write_xlsx(owa.p, "Supplementary Tables/oneway_anova_Drugs_vs_Cell_Type.xlsx")
```

```{r}
pdf("output/UMAP_Drug_response_NPM1_AUC.pdf")
lapply(names(meta.BEAT.drug), function(drug.name){
  
  

  drug.df <- meta.BEAT.drug[[drug.name]]
  temp.plot <- merge(plot.df[,c("Sample","Other.Sample", "UMAP.1", "UMAP.2", "cluster", "NPM1")], 
                     drug.df, by.x = "Other.Sample", by.y = "lab_id", all.x = T)
  
  temp.plot <- temp.plot %>% mutate(AUC.norm = normalize(auc, na.rm = T))
  temp.plot <- temp.plot %>% mutate(cluster.refined = ifelse(cluster %in% c("CD14 Mono", "HSC", "EMP", "T Cells", "GMP"), cluster, "Others"))
  
  plot.drug<- ggplot(temp.plot[sample(1:nrow(temp.plot)),] , aes(UMAP.1, UMAP.2, color = AUC.norm)) +
    geom_point(size = 2, data = temp.plot[is.na(temp.plot$auc),], color = "gray") + 
    geom_point(size = 2, data = temp.plot[ifelse(!is.na(temp.plot$auc),temp.plot$NPM1 > 0,FALSE), ]) +
    theme.gg + ggtitle(drug.name) + 
    coord_fixed(ratio = 1) + labs(color = "AUC") + 
    scale_color_viridis_c(option = "G", direction = -1)
    
  
  plot.drug.cluster<- ggplot(temp.plot %>% filter(NPM1 > 0) , aes(cluster.refined, AUC.norm, fill = cluster.refined)) +
    geom_boxplot(outlier.shape = NA) + 
    geom_jitter(size = 1, width = 0.1) +
    coord_flip() + 
    theme.gg + ggtitle(drug.name) + labs(color = "AUC") +
    scale_fill_manual(values = cell.type.colors) + 
    theme(legend.position = "none") + xlab("")
  
  p.val <- summary(aov(auc ~ cluster.refined, temp.plot))[[1]]$`Pr(>F)`[1]
  
  plot.drug | plot.drug.cluster + labs(caption = sprintf('anova p-value: %.5f', p.val)) 
  
}) 

dev.off()


# one way anova
owa.p <- lapply(names(meta.BEAT.drug), function(drug.name){
  
  

  drug.df <- meta.BEAT.drug[[drug.name]]
  temp.plot <- merge(plot.df[,c("Sample","Other.Sample", "UMAP.1", "UMAP.2", "cluster", "NPM1")], 
                     drug.df, by.x = "Other.Sample", by.y = "lab_id", all.x = T)
  
  temp.plot <- temp.plot %>% mutate(AUC.norm = normalize(auc, na.rm = T))
  temp.plot <- temp.plot %>% mutate(cluster.refined = ifelse(cluster %in% c("CD14 Mono", "HSC", "EMP", "T Cells", "GMP"), cluster, "Others"))
  
  p = summary(aov(auc ~ cluster.refined, temp.plot %>% filter(NPM1 > 0)))[[1]]$`Pr(>F)`[1]
  return(cbind(Drug = drug.name, p = sprintf("%.2e", p ), logp = -log(p)))
  
}) %>% do.call(rbind, .) %>% as.data.frame

owa.p$logp <- as.numeric(owa.p$logp)
owa.p<- owa.p[order(-owa.p$logp),]
owa.p$adj.p <- p.adjust(owa.p$p, method = "BH")

writexl::write_xlsx(owa.p, "Supplementary Tables/oneway_anova_Drugs_vs_Cell_Type_NPM1.xlsx")
```

# Random Forest Survival

```{r}
library(randomForestSRC)
set.seed(5)
pdf("output/RF_Survival_LOCOV.pdf")
for(x in c("TCGA", "BEAT", "TARGET")){
  
  cat(x)
  
  df.rfsrc <- plot.df[plot.df$Study != x & !duplicated(plot.df$Patient), c("overal_survival", "Status", cell.type.order)] 
  
  # cell.types <- cell.type.order[!colSums(df.rfsrc[,cell.type.order] > 0) <= 20 & 
  #                  colMeans(df.rfsrc[, cell.type.order]) > 0.05]
  # 
  # cell.types <- cell.type.order
  #   
  # df.rfsrc <- df.rfsrc[, c("overal_survival", "Status", cell.types)]


  model.rfsrc<- rfsrc(Surv(overal_survival,  Status) ~ . , data=df.rfsrc, ntree = 1000, importance = T)

  plot(model.rfsrc)
  plot.df$RFSRC.pred <- predict(model.rfsrc, newdata = plot.df[, cell.type.order, drop = F])$predicted
  
  par(pty="s", mfrow = c(1,1))
  plot(scale(plot.df[plot.df$Study == x,]$Stemness, center = T), scale(plot.df[plot.df$Study == x,]$RFSRC.pred, center = T), 
       pch = 16, ylab = "RF Score", xlab = "Stemness Score", xlim = c(-3,3), ylim = c(-3,3))
  
  new.df <- plot.df[plot.df$Study == x,]
  model <- survfit(Surv(overal_survival,  Status) ~ RFSRC.pred > median(RFSRC.pred, na.rm = T), 
                     data=new.df)
  
  
  
  p <- ggsurvplot(model, 
        conf.int = F,
        legend.labs = c("RF Score < Median", "RF Score > Median"),
        palette =  c("#085ea8", "#d60e00"),
        pval = sprintf("p = %.3f",surv_pvalue(model, method = "1")$pval),
        title = x,
        # # Add risk table
        risk.table = TRUE, censor.size = 3,
        tables.height = 0.25, legend = "none",
        tables.theme = theme_cleantable(),
        
        ggtheme = theme_bw(base_size = 10))

  print(p)
}
dev.off()

```

```{r Predict HSC/EMP}
ggsurvplot(survfit(Surv(overal_survival,  Status) ~ APS > median(APS, na.rm = T) , data=plot.df %>% filter(NPM1 ==1)), 
           pval = T,conf.int = F, risk.table = T)
```

```{r timeROC}
ggsurvplot(survfit(Surv(overal_survival,  Status) ~ ELN , data=plot.df[plot.df$Study == "TARGET",]), pval = T,conf.int = F, risk.table = F)
```

```{r Predicting APS scores}
temp <- plot.df

binary.celltypes <- cell.type.order[colMeans(temp[temp$Study == "BEAT", cell.type.order]) < 0.05]
temp[,binary.celltypes] <- ifelse(temp[,binary.celltypes] > 0, 1, 0)

coef.mat <- matrix(0, length(cell.type.order), 100)
rownames(coef.mat) <- cell.type.order


train.idx <- temp$Study == "BEAT" & !duplicated(temp$Patient) & temp$Recurrence == "No" 


coef.mat <- matrix(0, length(cell.type.order), 100)
rownames(coef.mat) <- cell.type.order
for(i in 1:ncol(coef.mat)){
  set.seed(i)
  cv.fit <- cv.glmnet(x = as.matrix(temp[train.idx,cell.type.order]), 
                    y = temp[train.idx, "APS"], nfolds = 10, intercept = F)
  coef.mat[,i] <- as.matrix(coef(cv.fit, s = cv.fit$lambda.min))[-1,]
}

lasso.coefs <- rowMeans(coef.mat)[rowMeans(coef.mat) != 0 & rowSums(abs(coef.mat) > 0) > 90]

lasso.coefs




idx.test <- ifelse(!duplicated(temp$Patient), 
                   ifelse(temp$Study == "BEAT", 
                          plot.df$Recurrence == "No", TRUE),
                   FALSE)

temp$APS.preds.lasso <- NA
temp[idx.test,]$APS.preds.lasso <- predict(cv.fit, newx = as.matrix(temp[idx.test,cell.type.order]), 
                                           s = cv.fit$lambda.min)

temp$APS.preds.lasso <- NA
temp[idx.test,]$APS.preds.lasso <- rowSums(sweep(temp[idx.test,names(lasso.coefs)], 2, lasso.coefs, FUN = "*"))

library(ggpmisc)
formula <- y ~ x
ggplot(temp, aes(APS, APS.preds.lasso)) + geom_point() + facet_wrap(~ Study) + 
  geom_smooth(method = "lm", formula = formula, se = F, color ="black") +
  stat_poly_eq(aes(label = paste(..rr.label.., sep = "~~~")), 
               label.x.npc = "left", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 3) + theme.gg

plot.ICS <- ggplot(temp[sample(1:nrow(temp)),] , aes(UMAP.1, UMAP.2, color = APS.preds.lasso)) +
  geom_point(size = 2, data = temp[ is.na(temp$APS.preds.lasso),], color = "gray") + 
  geom_point(size = 2, data = temp[!is.na(temp$APS.preds.lasso),]) +
  theme_light() + 
  coord_fixed(ratio = 1) +
  scale_color_viridis_c(option = "B")

plot.ICS
```

```{r Predicting APS scores}
temp <- plot.df

binary.celltypes <- cell.type.order[colMeans(temp[, cell.type.order]) < 0.05]
temp[,binary.celltypes] <- ifelse(temp[,binary.celltypes] > 0, 1, 0)

setdiff(cell.type.order, binary.celltypes)

fit <- lm(APS ~  . + 0, 
          temp[temp$Study == "BEAT" & !duplicated(plot.df$Patient) & plot.df$Recurrence == "No",
               c("APS", cell.type.order)])
summary(fit)

library(MASS)
fit <- stepAIC(fit)

temp$APS.preds.lm <- predict(fit, newdata = temp[,cell.type.order])


ggplot(temp, aes(APS, APS.preds.lm)) + geom_point() + facet_wrap(~ Study) + 
  geom_smooth(method = "lm", formula = formula, se = F, color ="black") +
  stat_poly_eq(aes(label = paste(..rr.label.., sep = "~~~")), 
               label.x.npc = "left", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 3) + theme.gg


plot.APS.preds<- ggplot(temp[sample(1:nrow(temp)),] , aes(UMAP.1, UMAP.2, color = APS.preds.lm)) +
  geom_point(size = 2) +
  theme_light() + 
  coord_fixed(ratio = 1) + 
  scale_color_viridis_c(option = "B", na.value = NA)
plot.APS.preds


ggplot(temp, aes(GMP, APS)) + geom_point()

ggsurvplot(survfit(Surv(overal_survival,  Status) ~ APS.preds.lm > median(APS.preds.lm) , data=temp[temp$Study == "TCGA",]), 
           pval = T,conf.int = F, risk.table = F)
ggsurvplot(survfit(Surv(overal_survival,  Status) ~ APS.preds.lm > median(APS.preds.lm) , data=temp[temp$Study == "BEAT",]), 
           pval = T,conf.int = F, risk.table = F)
ggsurvplot(survfit(Surv(overal_survival,  Status) ~ APS.preds.lm > median(APS.preds.lm) , data=temp[temp$Study == "TARGET",]), 
           pval = T,conf.int = F, risk.table = F)

```

```{r}
idx <- intersect(intersect(which(complete.cases(plot.df[, c("overal_survival", "Status")])), 
                           which(plot.df[, c("overal_survival")] != 0)), 
                 which(plot.df$Study == "TCGA"))

# store it in a temp
temp <- plot.df

binary.celltypes <- cell.type.order[colMeans(temp[temp$Study == "TCGA", cell.type.order]) < 0.01]

temp[,binary.celltypes] <- ifelse(temp[,binary.celltypes] > 0, 1, 0)


x <- temp[idx, cell.type.order]
y <- Surv(temp[idx,"overal_survival"], temp[idx,"Status"])




coef.mat <- matrix(0, length(cell.type.order), 100)
rownames(coef.mat) <- cell.type.order
for(i in 1:ncol(coef.mat)){
  set.seed(i)
  suppressWarnings(cvfit <- cv.glmnet(as.matrix(x), y, family = "cox", nfolds = 10, intercept = F, weights = rowSums(x > 0)))
  coef.mat[,i] <- as.matrix(coef(cvfit, s = cvfit$lambda.min))
}

lasso.coefs <- rowMeans(coef.mat)[rowMeans(coef.mat) != 0 & rowSums(abs(coef.mat) > 0 ) > 90]

lasso.coefs

temp$lasso.scores <- rowSums(sweep(temp[,names(lasso.coefs)], 2, lasso.coefs, FUN = "*"))

ggsurvplot(survfit(Surv(overal_survival,  Status) ~ lasso.scores >  median(lasso.scores), data=temp[temp$Study == "BEAT",]), pval = T,conf.int = F, risk.table = F)

ggsurvplot(survfit(Surv(overal_survival,  Status) ~ lasso.scores > median(lasso.scores) , data=temp[temp$Study == "TCGA",]), 
           pval = T,conf.int = F, risk.table = F)

ggsurvplot(survfit(Surv(overal_survival,  Status) ~ lasso.scores > median(lasso.scores), data=temp[temp$Study == "TARGET",]), 
           pval = T,conf.int = F, risk.table = F)



ggplot(temp, aes(UMAP.1, UMAP.2, color = lasso.scores)) +
  geom_point(size = 2) +
  theme_light() + 
  coord_fixed(ratio = 1) + 
  scale_color_viridis_c()

```

```{r NPM1}


df.npm1 <- plot.df %>% filter(NPM1 == "Yes")


df.npm1$NPM1.clusters <- ifelse(df.npm1$cluster %in% c("GMP", "CD14 Mono", "CD16 Mono"), "GMP & Monocytes", "Others")
model.npm1 <- survfit(Surv(overal_survival,  Status) ~ NPM1.clusters , data=df.npm1)

ggsurvplot(model.npm1, 
          pval = TRUE, # Change legends: title & labels
          legend.title = "Cluster", conf.int = F,
          legend.labs = gsub("NPM1.clusters=", "", names(model.npm1$strata)),
          
           
           # # Add risk table
           risk.table = TRUE, censor.size = 3,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))

ggplot(data=df.npm1, aes(Sex, Age)) +geom_violin() + geom_boxplot(width = 0.2) + geom_jitter(width = 0.1) + theme.gg + stat_compare_means(method = "t.test")

```

```{r}
# redefine groups <15 patients as 'Others'
meta.train <- meta.train %>%  dplyr::add_count(cluster)
meta.train$cluster.refined <- ifelse(meta.train$n > 50, meta.train$cluster, NA)
meta.train$Age.categorical <- ifelse(meta.train$Age >= 55, ">=55", "<55")


km.train.cluster.refined <- survfit(Surv(overal_survival, Status) ~ cluster.refined, 
                        data = meta.train[!duplicated(meta.train$Patient),])

km.train.cluster.refined
plot.train.cluster.refined <- ggsurvplot(km.train.cluster.refined,   
          pval = TRUE, # Change legends: title & labels
          legend.title = "Cluster", conf.int = F,
          legend.labs = gsub("cluster.refined=", "", names(km.train.cluster.refined$strata)),
           
           palette = cell.type.colors, 
           
           # # Add risk table
           risk.table = TRUE, censor.size = 3,
           tables.height = 0.25, legend = "none",
           tables.theme = theme_cleantable(),
           
           ggtheme = theme_bw(base_size = 12))
plot.train.cluster.refined
```

```{r}
library(ggExtra)



plotPerc <- function(x){
  p1 <- ggplot(temp.train, aes(Age, .data[[x]] * 100, color= Study)) + 
    geom_point() + 
    geom_smooth(method = "lm", se = F, color = cell.type.colors[x]) + 
    ylab(paste0("Deconvoluted ", x, "%")) + stat_cor() + 
    theme_light() + scale_color_manual(values = ann.colors$Study) + theme(legend.position = "bottom")
  
  ggMarginal(p1, groupColour = T, groupFill = T, margins = "both")
}

pdf("output/Age_vs_cell_types.pdf")
lapply(colnames(deconv.train), FUN = function(y) plotPerc(x=y))
dev.off()
```

```{r Marieke Stuff}

# df.NPM1 <- deconv.LUMC[rownames(meta.train[meta.train$Study == "LUMC" & meta.train$NPM1 == 1,]),]
# 
# clusters.NPM1 <- apply(df.NPM1, 1, function(x){names(x)[which(x == max(x))]})
# new.order.NPM1 <- patientOrder(df.NPM1, clusters.NPM1, cell.type.order)
# 
# df.NPM1 <- df.NPM1 %>% melt() %>% as.data.frame()
# df.NPM1$Var1 <- factor(df.NPM1$Var1, levels = rev(new.order.NPM1))
# 
# pdf("output/NPM1_LUMC_IC.pdf")
#  df.NPM1 %>% ggplot(aes(Var1, value * 100, fill = Var2)) + geom_bar(stat = "identity") + theme.gg + scale_fill_manual(values = cell.type.colors[1:22]) + coord_flip() + ylab("Cell Type %") + xlab("") + labs(fill = "Cell Types") + ggtitle("Immune compositions of NPM1 samples")
#  dev.off()

```

```{r}
# Can we predict how a patient will respond to a specific drug 
# given his/her immune profile?
meta.BEAT.drug <- readxl::read_excel("data/Studies_Latest/BEAT-AML/BEAT_Paper_meta.xlsx", sheet = "Table S10-Drug Responses")

temp.plot <- merge(plot.df[,c("Sample","Other.Sample")] , 
                   meta.BEAT.drug, by.x = "Other.Sample", by.y = "lab_id", all.x = T)

temp.plot <- temp.plot[!is.na(temp.plot$auc) & temp.plot$Sample %in% rownames(deconv.BEAT),]

# D_ij: samples vs drugs (Patients' drug responses)
D <- reshape2::acast(Sample ~ inhibitor,value.var = 'auc' ,data = temp.plot)

normalize <- function(x, ...) {
    return((x - min(x, ...)) /(max(x, ...) - min(x, ...)))
}

D <- apply(D, 2, function(x) normalize(x, na.rm = T))

# S_ij: samples vs cell types
S <- deconv.BEAT[rownames(D),]


# as a dummy variable
predictedRes <- D

# # over each drug
# for (k in 1:ncol(D)){
#   
#   colnames(D)[k]
#   # loocv 
#   for (i in 1:nrow(D)){
#     cat("\n", i)
#     Xtest <- S[i,,drop=F]
#     ytest <- D[i,k]    
#     
#     if(is.na(ytest)){
#       predictedRes[i,k] <- NA
#     } else{
      # Xtrain <- S[-i,]
      # ytrain <- D[-i,k]
      # 
      # idx <- !is.na(ytrain)
      # 
      # Xtrain <- Xtrain[idx,]
      # ytrain <- ytrain[idx]
# 
#       predictedRes[i,k] <- predict(randomForest::randomForest(x = Xtrain, y=ytrain), newdata = Xtest)
#     }
#     
#     
#   }
# }

load("output/DrugResponsePreds.RDS")

writexl::write_xlsx(as.data.frame(cbind(Sample = rownames(D),D)), "Supplementary Tables/Normalized_Drug_Values.xlsx")
writexl::write_xlsx(as.data.frame(cbind(Sample =rownames(predictedRes),predictedRes)), "Supplementary Tables/Drug_Predictions.xlsx")
writexl::write_xlsx(predCor, "Supplementary Tables/DrugPredictionCorrelations.xlsx")

# dummy var
predCor <- D[1,]
# 
# pdf("output/Drug_Response_Prediction_RF_LOOCV_Scatter_Primary.pdf")
# formula <- y ~ x
# for(k in 1:ncol(D)){
#   
#   temp <- merge(meta.train, cbind(D[,k], predictedRes[,k]), by.x = "Sample", by.y = 0)
#   
#   predCor[k] <- cor(temp$V1,temp$V2, use="complete.obs", method = "spearman")
#   
#   p <- ggplot(temp, aes(V1, V2, color = `Primary Diagnosis`)) + geom_point() + 
#     theme.gg + xlim(c(0,1)) + ylim(c(0,1)) + coord_fixed() + ggtitle(colnames(D)[k]) + 
#     annotate(geom="text", x=0.15, y=0.95, parse = T, label=sprintf("~ rho == %.3f", predCor[k])) +
#     scale_color_manual(values = DiscretePalette(20)) + xlab("AUC") + ylab("Predicted AUC")
#   
#   print(p)
# }
# dev.off()

temp <- merge(meta.train, cbind(D[,"Venetoclax"], predictedRes[,"Venetoclax"]), by.x = "Sample", by.y = 0)
corr <- cor(temp$V1,temp$V2, use="complete.obs", method = "spearman")


temp$`Primary Diagnosis` <- stringr::str_wrap(temp$`Primary Diagnosis`, width = 20)
temp <- temp %>% filter(!is.na(V2))%>% add_count(`Primary Diagnosis`)

pdf("output/Venetoclax_Response_Primary_Diagnosis.pdf", width = 9, height = 9)
formula <- y ~ x
ggplot(temp, aes(V1, V2)) + 
  geom_point(data=dplyr::select(temp, -`Primary Diagnosis`), color = "gray") + 
  geom_point(size=3,pch=21, aes(fill=`Primary Diagnosis`)) + 
  theme.gg + xlim(c(0,1)) + ylim(c(0,1)) + ggtitle("") + 
  geom_smooth(data = subset(temp, n>5), method = "lm", se = F, size =0.5, color = "black")+ 
  stat_poly_eq(aes(label = ifelse(n > 5, paste(..rr.label.., ..p.value.label.., sep = "~~~"), "")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 4 ) +
  scale_color_manual(values = DiscretePalette(20)) + xlab("Normalized AUC") + ylab("Predicted AUC") + 
  facet_wrap(~ `Primary Diagnosis`) + theme_bw(base_size = 12) +  theme(legend.position = "none")
dev.off()


pdf("output/Venetoclax_Response_Assignment.pdf", width = 8, height = 8)
ggplot(temp, aes(V1, V2)) + 
  geom_point(size=2, aes(color=cluster)) + coord_fixed() +
  theme.gg + xlim(c(0,1)) + ylim(c(0,1)) + ggtitle("Venetoclax") +
  scale_color_manual(values =cell.type.colors[1:22]) + xlab("AUC") + ylab("Predicted AUC") + 
  theme(legend.position = "bottom") + labs(color = "Assignment") + ggtitle("")
dev.off()





pdf("output/Drug_Response_Prediction_RF_LOOCV_Scatter.pdf")
formula <- y ~ x
for(k in 1:ncol(D)){
  
  temp <- merge(meta.train, cbind(D[,k], predictedRes[,k]), by.x = "Sample", by.y = 0)
  
  predCor[k] <- cor(temp$V1,temp$V2, use="complete.obs", method = "spearman")
  
  p <- ggplot(temp, aes(V1, V2)) + geom_point(size = 2) + 
    xlim(c(0,1)) + ylim(c(0,1)) + coord_fixed() + ggtitle(colnames(D)[k]) + 
    annotate(geom="text", x=0.15, y=0.95, parse = T, size = 5, label=sprintf("~ rho == %.3f", predCor[k])) +
    xlab("Normalized AUC") + ylab("Predicted AUC") + theme_bw(base_size = 16)
  
  print(p)
}
dev.off()


predCor <- tibble::enframe(predCor)

pdf("output/Drug_Response_Prediction_RF_LOOCV_Rho.pdf", width = 10, height = 5)
predCor %>% 
  ggplot(aes(reorder(name, -value), value)) + 
  geom_bar(stat = "identity") + 
  theme_bw(base_size =  8) + 
   xlab("") + scale_x_discrete(guide = guide_axis(angle = 90)) + 
  ylim(c(-1, 1)) + ylab("Spearman's Rho") + 
  labs(subtitle = "Leave one out cross validation within each drug")
dev.off()
```

```{r BEAT Venetoclax Response vs CD14 Mono %}
meta.train.BEAT <- plot.df[plot.df$Study == "BEAT",]


contrast.BEAT<- ifelse(meta.train.BEAT$cluster == "CD14 Mono", "CD14Mono", "Others")

res.BEAT <- cinaR::cinaR(matrix = cbind(rownames(count.BEAT[,meta.train.BEAT$Sample]), count.BEAT[,meta.train.BEAT$Sample]), 
                              contrasts = contrast.BEAT, 
                              experiment.type = "RNA-Seq")


cp.BEAT <- res.BEAT$DA.results$cp

df.BCL2 <- data.frame(BCL2 = cp.BEAT["BCL2",], meta.train.BEAT)

drug.res.BCL2 <- temp.plot [temp.plot$inhibitor == "Venetoclax" & temp.plot$Sample %in% meta.train.BEAT$Sample, ]

df.BCL2 <- merge(df.BCL2, drug.res.BCL2, by = "Sample")
df.BCL2 <- merge(df.BCL2, cbind(Sample = rownames(predictedRes),predictedRes), by = "Sample")

pdf("output/Venetoclax_BCL2_vs_AUC.pdf", width = 8, height = 6)
df.BCL2 %>% ggplot(aes(normalize(auc), BCL2, color = CD14.Mono)) + geom_point(size = 3)  + xlab("Venetoclax Resistance\n(Normalized AUC)") + ylab("BCL2 Expression\n(cpm)")+ scale_color_gradient(low = "gray" ,high = "#b00000") + labs(color = "CD14 Mono") + theme_bw(base_size = 12)
dev.off()
```


```{r Venetoclax Response with LM}
colnames(df.BCL2)

df.BCL2.filtered <- df.BCL2[,c(colnames(df.BCL2)[c(2, 7:27, 31:38, 96)],"auc" ,"NPM1", "FLT3")]

df.BCL2.filtered <- df.BCL2.filtered[, apply(df.BCL2.filtered, 2, function(x){length(unique(x)) > 1}) %>% as.vector]

df.summaries <- df.BCL2.filtered %>% 
  dplyr::select(-auc) %>% 
  purrr::map(~lm(normalize(df.BCL2.filtered$auc) ~ .x, data = df.BCL2.filtered)) %>% 
  purrr::map(summary) %>% 
  purrr::map(c("coefficients"))
  
pvals <- sapply(names(df.summaries), function(x){
  res <- df.summaries[[x]][-1,4]
  if(!is.null(names(res))){
    names(res) <- gsub(".x", "", names(res))
  }
  return(res)
}) %>% unlist
  

coefs <- sapply(names(df.summaries), function(x){
  res <- df.summaries[[x]][-1,1]
  if(!is.null(names(res))){
    names(res) <- gsub(".x", "", names(res))
  }
  return(res)
}) %>% unlist


df.BCL2.plot <- data.frame(Attribute = names(pvals), p = pvals, logP= -log10(pvals), Coefficient = coefs)
df.BCL2.plot$adjP <- p.adjust(df.BCL2.plot$p, method = "BH")
df.BCL2.plot$log.adjP <- -log10(df.BCL2.plot$adjP)

writexl::write_xlsx(df.BCL2.plot, "Supplementary Tables/Univariate_Venetoclax_Resistance.xlsx")

pdf("output/Venetoclax_Response_Attributes.pdf", width = 7, height = 6)
ggplot(df.BCL2.plot, aes(Coefficient, log.adjP, label = ifelse(adjP < 0.01, Attribute, ""), size = log.adjP)) + 
  geom_point()  + 
  ggrepel::geom_text_repel(size = 5, force = 30) + theme_bw(base_size = 14) + theme(legend.position = "bottom") + 
  ylab("-log10(adj. p)") + geom_hline(yintercept = -log10(0.01), linetype='dotted', col = 'red') +
  annotate("text", x = -4.2, y = 2.5, label = "FDR = 0.01", color = "red", size = 4) + xlab("Effect Size") +
  labs(size = "-log10(adj. p)")

dev.off()
```


```{r Venetoclax Multivariate Response with LM}
df.BCL2.plot.multi <- 
  coef(summary(lm(normalize(auc) ~ CD14.Mono + BCL2 + Sex + Origin + Age + Primary.Diagnosis +Blast + NPM1 + FLT3 + ELN, data = df.BCL2.filtered))) %>% as.data.frame

colnames(df.BCL2.plot.multi)[4] <- "p"
df.BCL2.plot.multi$logP <- -log10(df.BCL2.plot.multi$p)

writexl::write_xlsx(df.BCL2.plot.multi, "Supplementary Tables/Multivariate_Venetoclax_Resistance.xlsx")

# remove intercept
df.BCL2.plot.multi <- df.BCL2.plot.multi[-1, ]
df.BCL2.plot.multi$Attribute <- rownames(df.BCL2.plot.multi)

pdf("output/Venetoclax_Response_Attributes_Multi.pdf", width = 7, height = 6)
ggplot(df.BCL2.plot.multi, aes(Estimate, logP, label = ifelse(p < 0.05, Attribute, ""), size = logP)) + 
  geom_point()  + 
  ggrepel::geom_text_repel(size = 5, force = 200) + theme_bw(base_size = 14) + theme(legend.position = "bottom") + 
  ylab("-log10(p)") + geom_hline(yintercept = -log10(0.01), linetype='dotted', col = 'red') +
  annotate("text", x = -0.28, y = 2.2, label = "p = 0.01", color = "red", size = 4) + xlab("Effect Size") +
  labs(size = "-log10(p)")

dev.off()


```


```{r}
df_corrected <- readRDS("data/Proteomics_LUMC.RDS")
cp <- df_corrected$matrix 
meta <- df_corrected$meta
rnaseqID <- strsplit(meta$rnaseqID, split = ";", fixed = T) %>% purrr::map(1) %>% unlist
rnaseqID[70] <- "102581-003-042"


rnaseqID <- strsplit(rnaseqID, split = "-", fixed = T) %>% lapply(function(x){
  if(!is.na(x)) {
    paste0("AML_", substr(x[2], 3,3 ),"_",x[3]) 
  } else {
    return(NA)
  }
}) %>% unlist


meta$Sample <- rnaseqID

# technical replicate
meta <- meta[!rownames(meta) %in% c("F1__132N", "F3__132N", "F1__127C"),]
meta <- meta[!is.na(meta$Sample) & meta$Tissue == "PRIM", ]


cp <- cp[, rownames(meta)]

mergedProteomics <- cbind(meta, BCL2.Proteomic = cp["P10415", ])

meta.LUMC$order <- 1:nrow(meta.LUMC)

meta.LUMC2 <- temp.LUMC

count.LUMC.norm <- cinaR::normalizeConsensus(count.LUMC, log.option = T)

meta.LUMC2$BCL2.Expression <- as.numeric(count.LUMC.norm["BCL2",])
meta.LUMC2$CD14.Mono <- deconv.LUMC[,"CD14 Mono"]
  
meta.LUMC2 <- merge(meta.LUMC2, mergedProteomics, by = "Sample", all.x = T)


library(ggpmisc)
formula <- y ~ x

cor(meta.LUMC2$BCL2.Expression, meta.LUMC2$BCL2.Proteomic, method = "spearman", use = "complete.obs")
cor(meta.LUMC2$CD14.Mono, meta.LUMC2$BCL2.Proteomic, method = "spearman", use = "complete.obs")

meta.LUMC2$`Primary Diagnosis` <- stringr::str_wrap(meta.LUMC2$`Primary Diagnosis`, width = 25)

p1 <- ggplot(meta.LUMC2 %>% filter(!is.na(BCL2.Proteomic)), aes(BCL2.Expression, BCL2.Proteomic)) + 
  geom_point(aes(color = mut),size = 3) + 
  geom_smooth(method = "lm", se = F, size =0.5, color = "black")+ 
  stat_poly_eq(aes(label = paste(..rr.label.., ..p.value.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 5) + theme_bw(base_size = 16) +
  xlab("BCL2 Expression (cpm)") + ylab("BCL2 Protein Abundance\n(cpm)") + labs(color = "Genetic Abnormalities") +
  scale_color_manual(values = rev(DiscretePalette(20)))

p2 <- ggplot(meta.LUMC2 %>% filter(!is.na(BCL2.Proteomic)), aes(CD14.Mono, BCL2.Proteomic)) + 
  geom_point(aes(color = mut),size = 3) +
  geom_smooth(method = "lm", se = F, size =0.5, color = "black")+ 
  stat_poly_eq(aes(label = paste(..rr.label.., ..p.value.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 5) + theme_bw(base_size = 16) +
  xlab("CD14+ Mono%") + ylab("BCL2 Protein Abundance\n(cpm)") + labs(color = "Genetic Abnormalities") +
  scale_color_manual(values = rev(DiscretePalette(20)))

pdf("output/BCL2_Expression_or_Monocyte_vs_Proteomic_PD.pdf", width = 16, height = 8)
print(ggarrange(p1, p2, common.legend = T, legend = "bottom"),newpage = F)
dev.off()


p3 <- ggplot(meta.LUMC2, aes(BCL2.Expression, BCL2.Proteomic)) + geom_point(aes(color = cluster), size = 3) + 
  geom_smooth(method = "lm", se = F, size =0.5, color = "black")+ 
  stat_poly_eq(aes(label = paste(..rr.label.., ..p.value.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 5) + theme_bw(base_size = 16) +
  xlab("BCL2 Expression (cpm)") + ylab("BCL2 Protein Abundance\n(cpm)") + labs(color = "Assignment") + 
  scale_color_manual(values = cell.type.colors[unique(meta.LUMC2[!is.na(meta.LUMC2$BCL2.Proteomic),]$cluster)])

p4 <- ggplot(meta.LUMC2, aes(CD14.Mono, BCL2.Proteomic)) + geom_point(aes(color = cluster), size = 3) + 
    geom_smooth(method = "lm", se = F, size =0.5, color = "black")+ 
   stat_poly_eq(aes(label = paste(..rr.label.., ..p.value.label..,
 sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 5) + theme_bw(base_size = 16) +
  xlab("CD14 Mono %") + ylab("BCL2 Protein Abundance\n(cpm)") + labs(color = "Assignment") + 
  scale_color_manual(values = cell.type.colors[unique(meta.LUMC2[!is.na(meta.LUMC2$BCL2.Proteomic),]$cluster)])

pdf("output/BCL2_Expression_or_Monocyte_vs_Proteomic.pdf", width = 16, height = 8)
print(ggarrange(p3, p4, common.legend = T, legend = "bottom"),newpage = F)
dev.off()
```

```{r}
df_corrected <- readRDS("data/Proteomics_LUMC.RDS")
cp <- df_corrected$matrix 
meta <- df_corrected$meta
rnaseqID <- strsplit(meta$rnaseqID, split = ";", fixed = T) %>% purrr::map(1) %>% unlist
rnaseqID[70] <- "102581-003-042"


rnaseqID <- strsplit(rnaseqID, split = "-", fixed = T) %>% lapply(function(x){
  if(!is.na(x)) {
    paste0("AML_", substr(x[2], 3,3 ),"_",x[3]) 
  } else {
    return(NA)
  }
}) %>% unlist


meta$Sample <- rnaseqID

# remove technical duplicates
meta <- meta[meta$mut == "NPM1", ]

# you're no npm1
meta <- meta[!rownames(meta) %in% c("F3__129N"),]
# technical replicate
meta <- meta[!rownames(meta) %in% c("F1__132N", "F3__132N", "F1__127C"),]
meta <- meta[!is.na(meta$Sample) & meta$Tissue == "PRIM", ]


cp <- cp[, rownames(meta)]

mergedProteomics <- cbind(meta, BCL2.Proteomic = cp["P10415", ])

meta.LUMC$order <- 1:nrow(meta.LUMC)

meta.LUMC2 <- temp.LUMC

count.LUMC.norm <- cinaR::normalizeConsensus(count.LUMC, log.option = T)

meta.LUMC2$BCL2.Expression <- as.numeric(count.LUMC.norm["BCL2",])
meta.LUMC2$CD14.Mono <- deconv.LUMC[,"CD14 Mono"]
  
meta.LUMC2 <- merge(meta.LUMC2, mergedProteomics, by = "Sample", all.x = T)


library(ggpmisc)
formula <- y ~ x

cor(meta.LUMC2$BCL2.Expression, meta.LUMC2$BCL2.Proteomic, method = "spearman", use = "complete.obs")
cor(meta.LUMC2$CD14.Mono, meta.LUMC2$BCL2.Proteomic, method = "spearman", use = "complete.obs")


p5 <- ggplot(meta.LUMC2, aes(BCL2.Expression, BCL2.Proteomic)) + geom_point(aes(color = cluster), size = 3) + 
  geom_smooth(method = "lm", se = F, size =0.5, color = "black")+ 
  stat_poly_eq(aes(label = paste(..rr.label.., ..p.value.label.., sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 5) + theme_bw(base_size = 16) +
  xlab("BCL2 Expression (cpm)") + ylab("BCL2 Protein Abundance\n(cpm)") + labs(color = "Assignment") + 
  scale_color_manual(values = cell.type.colors[unique(meta.LUMC2[!is.na(meta.LUMC2$BCL2.Proteomic),]$cluster)])



p6 <- ggplot(meta.LUMC2, aes(CD14.Mono, BCL2.Proteomic)) + geom_point(aes(color = cluster), size = 3) + 
    geom_smooth(method = "lm", se = F, size =0.5, color = "black")+ 
   stat_poly_eq(aes(label = paste(..rr.label.., ..p.value.label..,
 sep = "~~~")), 
               label.x.npc = "right", label.y.npc = 0.95,
               formula = formula, parse = TRUE, size = 5) + theme_bw(base_size = 16) +
  xlab("CD14 Mono %") + ylab("BCL2 Protein Abundance\n(cpm)") + labs(color = "Assignment") + 
  scale_color_manual(values = cell.type.colors[unique(meta.LUMC2[!is.na(meta.LUMC2$BCL2.Proteomic),]$cluster)])

pdf("output/BCL2_Expression_or_Monocyte_vs_Proteomic_NPM1.pdf", width = 16, height = 8)
print(ggarrange(p5, p6, common.legend = T, legend = "bottom"),newpage = F)
dev.off()

```

```{r}

mat.meta <- meta.train[new.order.train,]

idx <- rowSds(as.matrix(mat.meta[,common.mutations]), na.rm = T)
idx[is.na(idx)] <- 0

idx <- idx > 0 

mat <- mat.meta[idx, common.mutations]
mat[mat > 0] <- 1
mat[is.na(mat)] <- 0

mat.hm <- mat.meta[idx, "cluster", drop = F]

mat.vec <- mat.hm$cluster
names(mat.vec) <- rownames(mat.hm)
pheatmap(t(mat), annotation_col = mat.hm, show_colnames = F, cluster_rows = T, cluster_cols = F,
         annotation_colors = ann.colors, border_color = NA, 
        gaps_col = cumsum(table(mat.vec)[unique(mat.vec)]))

```
```{r}
colSums(meta.train[,common.mutations],na.rm = T) %>% sort(decreasing = T)


meta.train %>% 
  dplyr::filter(FLT3 > 0 & !duplicated(Patient) & !is.na(Age)) %>% 
  ggplot(aes(Age, color = Sex)) + geom_density()
```

